{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## In PMI cm03\n",
    "### Environment\n",
    "```bash\n",
    "source activate scanpy_1.9.3\n",
    "ipython --profile=ak1\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from anndata import AnnData\n",
    "import anndata\n",
    "from scipy import sparse, io\n",
    "import scipy\n",
    "import pandas as pd\n",
    "import scipy.io\n",
    "import os\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import matplotlib.colors\n",
    "from matplotlib.ticker import StrMethodFormatter\n",
    "matplotlib.use('TkAgg')\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import math\n",
    "import scanpy.external as sce\n",
    "#import scrublet as scr\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.formula.api import ols\n",
    "from statsmodels.stats.multicomp import pairwise_tukeyhsd\n",
    "from scipy.stats import chi2_contingency\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats import zscore\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "sns.set(font=\"arial\", font_scale=1.15, style='ticks')\n",
    "plt.rcParams['figure.figsize'] = (6,6)\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n",
    "sc.settings.verbosity = 3\n",
    "\n",
    "%autoindent\n",
    "%matplotlib\n",
    "\n",
    "cmap = matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#104e8b\", \"#ffdab9\", \"#8b0a50\"])\n",
    "sample_palette = {'AK1':'#FF0000', 'iPSC':'#154360','NPC':'#229954'}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Initial Diagnosis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ak1 = pd.read_table(\"AK1_collapsed_classification.filtered_lite_classification_saturation.txt\")\n",
    "ak1_melt = pd.melt(ak1, id_vars=['reads'], value_vars=['unique_genes', 'unique_isoforms', 'unique_genes_known', 'unique_isoforms_known'])\n",
    "sns.lineplot(data=ak1_melt, x='reads', y='value', hue='variable')\n",
    "sns.plt.savefig('lineplot.pdf')\n",
    "\n",
    "ipsc = pd.read_table(\"iPSC_collapsed_classification.filtered_lite_classification_saturation.txt\")\n",
    "ipsc_melt = pd.melt(ipsc, id_vars=['reads'], value_vars=['unique_genes', 'unique_isoforms', 'unique_genes_known', 'unique_isoforms_known'])\n",
    "sns.lineplot(data=ipsc_melt, x='reads', y='value', hue='variable')\n",
    "\n",
    "npc = pd.read_table(\"NPC_collapsed_classification.filtered_lite_classification_saturation.txt\")\n",
    "npc_melt = pd.melt(npc, id_vars=['reads'], value_vars=['unique_genes', 'unique_isoforms', 'unique_genes_known', 'unique_isoforms_known'])\n",
    "sns.lineplot(data=npc_melt, x='reads', y='value', hue='variable')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Single-Cell Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ak1 = sc.read_10x_mtx('/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/AK1/genes_seurat', cache=True)\n",
    "ipsc = sc.read_10x_mtx('/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/iPSC/genes_seurat', cache=True)\n",
    "npc = sc.read_10x_mtx('/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/NPC/genes_seurat', cache=True)\n",
    "\n",
    "ak1_iso = sc.read_10x_mtx('/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/AK1/isoforms_seurat', cache=True)\n",
    "ipsc_iso = sc.read_10x_mtx('/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/iPSC/isoforms_seurat', cache=True)\n",
    "npc_iso = sc.read_10x_mtx('/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/NPC/isoforms_seurat', cache=True)\n",
    "\n",
    "ak1.write(filename=\"AK1_genes.h5ad\")\n",
    "ipsc.write(filename=\"iPSC_genes.h5ad\")\n",
    "npc.write(filename=\"NPC_genes.h5ad\")\n",
    "ak1_iso.write(filename=\"AK1_isoforms.h5ad\")\n",
    "ipsc_iso.write(filename=\"iPSC_isoforms.h5ad\")\n",
    "npc_iso.write(filename=\"NPC_isoforms.h5ad\")\n",
    "\n",
    "'''\n",
    "knee = np.sort((np.array(ak1.X.sum(axis=1))).flatten())[::-1] # UMI count for each cell (axis=1)\n",
    "cell_set = np.arange(len(knee))\n",
    "cutoff = 200\n",
    "num_cells = cell_set[knee > cutoff][::-1][0]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 7))\n",
    "\n",
    "ax.loglog(knee, cell_set, lw=5, color=\"g\")\n",
    "ax.axvline(x=num_cells, lw=3, color=\"k\")\n",
    "ax.axhline(y=cutoff, lw=3, color=\"k\")\n",
    "ax.set_xlabel(\"Set of Barcodes\")\n",
    "ax.set_ylabel(\"UMI counts\")\n",
    "ax.grid(True, which = \"both\")\n",
    "\n",
    "print(f\"{num_cells:,.0f} cells passed the {cutoff} UMI threshold\")\n",
    "#9,364 cells passed the 200 UMI threshold (2022-10-22)\n",
    "'''\n",
    "\n",
    "ak1 = sc.read_10x_mtx('/mnt/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/New_2024/NEW/AK1/genes_seurat', cache=True)\n",
    "ipsc = sc.read_10x_mtx('/mnt/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/New_2024/NEW/iPSC/genes_seurat', cache=True)\n",
    "npc = sc.read_10x_mtx('/mnt/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/New_2024/NEW/NPC/genes_seurat', cache=True)\n",
    "ak1.var_names_make_unique()\n",
    "ipsc.var_names_make_unique()\n",
    "npc.var_names_make_unique()\n",
    "\n",
    "ak1_iso = sc.read_10x_mtx('/mnt/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/New_2024/NEW/AK1/isoforms_seurat', cache=True)\n",
    "ipsc_iso = sc.read_10x_mtx('/mnt/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/New_2024/NEW/iPSC/isoforms_seurat', cache=True)\n",
    "npc_iso = sc.read_10x_mtx('/mnt/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/New_2024/NEW/NPC/isoforms_seurat', cache=True)\n",
    "ak1_iso.var_names_make_unique()\n",
    "ipsc_iso.var_names_make_unique()\n",
    "npc_iso.var_names_make_unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ak1.var['mt'] = ak1.var_names.str.startswith('MT-')\n",
    "sc.pp.calculate_qc_metrics(ak1, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "ipsc.var['mt'] = ipsc.var_names.str.startswith('MT-')\n",
    "sc.pp.calculate_qc_metrics(ipsc, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "npc.var['mt'] = npc.var_names.str.startswith('MT-')\n",
    "sc.pp.calculate_qc_metrics(npc, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "ak1_iso.var['mt'] = ak1_iso.var_names.str.startswith('MT-')\n",
    "sc.pp.calculate_qc_metrics(ak1_iso, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "ipsc_iso.var['mt'] = ipsc_iso.var_names.str.startswith('MT-')\n",
    "sc.pp.calculate_qc_metrics(ipsc_iso, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "npc_iso.var['mt'] = npc_iso.var_names.str.startswith('MT-')\n",
    "sc.pp.calculate_qc_metrics(npc_iso, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene-level\n",
    "integrated = AnnData.concatenate(ak1, ipsc, npc, join='outer', batch_categories = ['AK1', 'iPSC', 'NPC'], index_unique = '-')\n",
    "integrated.raw = integrated\n",
    "integrated.var['mt'] = integrated.var_names.str.startswith('MT-')\n",
    "sc.pp.calculate_qc_metrics(integrated, qc_vars=['mt'], percent_top=None, log1p=True, inplace=True)\n",
    "sc.pp.log1p(integrated)\n",
    "#integrated.write_h5ad(\"integrated.h5ad\")\n",
    "\n",
    "# Isoform-level\n",
    "integrated_iso = AnnData.concatenate(ak1_iso, ipsc_iso, npc_iso, join='outer', batch_categories = ['AK1', 'iPSC', 'NPC'], index_unique = '-')\n",
    "integrated_iso.raw = integrated_iso\n",
    "integrated_iso.var['mt'] = integrated_iso.var_names.str.startswith('MT-')\n",
    "sc.pp.calculate_qc_metrics(integrated_iso, qc_vars=['mt'], percent_top=None, log1p=True, inplace=True)\n",
    "sc.pp.log1p(integrated_iso)\n",
    "#integrated_iso.write_h5ad(\"integrated_iso.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Wilcoxon\n",
    "## Gene-level\n",
    "sc.tl.rank_genes_groups(integrated, groupby='batch', method='wilcoxon', corr_method='benjamini-hochberg', use_raw=False, pts=True, key_added='DEG_bw_batches_wilcoxon') # key_added=''\n",
    "dedf_ak1 = sc.get.rank_genes_groups_df(integrated, group='AK1', key='DEG_bw_batches_wilcoxon')\n",
    "dedf_ipsc = sc.get.rank_genes_groups_df(integrated, group='iPSC', key='DEG_bw_batches_wilcoxon')\n",
    "dedf_npc = sc.get.rank_genes_groups_df(integrated, group='NPC', key='DEG_bw_batches_wilcoxon')\n",
    "\n",
    "#dedf_ak1[(dedf_ak1['logfoldchanges'] > 2) & (dedf_ak1['pvals_adj'] < 10e-5)].sort_values(by='logfoldchanges', ascending=False)\n",
    "\n",
    "ak1_up_log2fc_2_genename = set(dedf_ak1[(dedf_ak1['logfoldchanges'] > 2) & (dedf_ak1['pvals_adj'] < 10e-5)]['names'].values)\n",
    "ak1_up_log2fc_1_genename = set(dedf_ak1[(dedf_ak1['logfoldchanges'] > 1) & (dedf_ak1['pvals_adj'] < 10e-5)]['names'].values)\n",
    "ak1_down_log2fc_2_genename = set(dedf_ak1[(dedf_ak1['logfoldchanges'] < -2) & (dedf_ak1['pvals_adj'] < 10e-5)]['names'].values)\n",
    "ak1_down_log2fc_1_genename = set(dedf_ak1[(dedf_ak1['logfoldchanges'] < -1) & (dedf_ak1['pvals_adj'] < 10e-5)]['names'].values)\n",
    "\n",
    "ak1_deg_log2fc_1_genename = ak1_up_log2fc_1_genename.union(ak1_down_log2fc_1_genename)\n",
    "total_genename = set(integrated.var.index)\n",
    "ak1_nondeg_log2fc_1_genename = total_genename - ak1_deg_log2fc_1_genename\n",
    "\n",
    "## Isoform-level\n",
    "sc.tl.rank_genes_groups(integrated_iso, groupby='batch', method='wilcoxon', corr_method='benjamini-hochberg', use_raw=False, pts=True, key_added='DEG_bw_batches_wilcoxon') # key_added=''\n",
    "dedf_ak1_iso = sc.get.rank_genes_groups_df(integrated_iso, group='AK1', key='DEG_bw_batches_wilcoxon')\n",
    "dedf_ipsc_iso = sc.get.rank_genes_groups_df(integrated_iso, group='iPSC', key='DEG_bw_batches_wilcoxon')\n",
    "dedf_npc_iso = sc.get.rank_genes_groups_df(integrated_iso, group='NPC', key='DEG_bw_batches_wilcoxon')\n",
    "\n",
    "#dedf_ak1_iso[(dedf_ak1_iso['logfoldchanges'] > 2) & (dedf_ak1_iso['pvals_adj'] < 10e-5)].sort_values(by='logfoldchanges', ascending=False)\n",
    "\n",
    "ak1_iso_up_log2fc_2_transname = set(dedf_ak1_iso[(dedf_ak1_iso['logfoldchanges'] > 2) & (dedf_ak1_iso['pvals_adj'] < 10e-5)]['names'].values)\n",
    "ak1_iso_up_log2fc_1_transname = set(dedf_ak1_iso[(dedf_ak1_iso['logfoldchanges'] > 1) & (dedf_ak1_iso['pvals_adj'] < 10e-5)]['names'].values)\n",
    "ak1_iso_down_log2fc_2_transname = set(dedf_ak1_iso[(dedf_ak1_iso['logfoldchanges'] < -2) & (dedf_ak1_iso['pvals_adj'] < 10e-5)]['names'].values)\n",
    "ak1_iso_down_log2fc_1_transname = set(dedf_ak1_iso[(dedf_ak1_iso['logfoldchanges'] < -1) & (dedf_ak1_iso['pvals_adj'] < 10e-5)]['names'].values)\n",
    "\n",
    "ak1_iso_dex_log2fc_1_transname = ak1_iso_up_log2fc_1_transname | ak1_iso_down_log2fc_1_transname\n",
    "\n",
    "ak1_iso_up_log2fc_1_genename = set(map(lambda x: '-'.join(x.split('-')[:-1]), ak1_iso_up_log2fc_1_transname))\n",
    "ak1_iso_down_log2fc_1_genename = set(map(lambda x: '-'.join(x.split('-')[:-1]), ak1_iso_down_log2fc_1_transname))\n",
    "ak1_iso_dex_log2fc_1_genename = ak1_iso_up_log2fc_1_genename | ak1_iso_down_log2fc_1_genename\n",
    "\n",
    "ak1_iso_dex_log2fc_2_transname = ak1_iso_up_log2fc_2_transname | ak1_iso_down_log2fc_2_transname\n",
    "ak1_iso_up_log2fc_2_genename = set(map(lambda x: '-'.join(x.split('-')[:-1]), ak1_iso_up_log2fc_2_transname))\n",
    "ak1_iso_down_log2fc_2_genename = set(map(lambda x: '-'.join(x.split('-')[:-1]), ak1_iso_down_log2fc_2_transname))\n",
    "ak1_iso_dex_log2fc_2_genename = ak1_iso_up_log2fc_2_genename | ak1_iso_down_log2fc_2_genename\n",
    "\n",
    "# Intersection of non-deg (Gene-level) and deg (Isoform-level)\n",
    "ak1_nodeg_log2fc_1_yesdex_log2fc_1_genename = ak1_nondeg_log2fc_1_genename & ak1_iso_dex_log2fc_1_genename #len(ak1_nodeg_yesdex_log2fc_1_genename) == 1365\n",
    "ak1_nodeg_log2fc_1_yesdex_log2fc_2_genename = ak1_nondeg_log2fc_1_genename & ak1_iso_dex_log2fc_2_genename #len(ak1_nodeg_yesdex_log2fc_1_genename) == 1365\n",
    "\n",
    "ak1_nodeg_log2fc_1_yesdex_log2fc_2_transname = set()\n",
    "for iso in dedf_ak1_iso['names'].values:\n",
    "    if '-'.join(iso.split('-')[:-1]) in ak1_nodeg_log2fc_1_yesdex_log2fc_2_genename:\n",
    "        ak1_nodeg_log2fc_1_yesdex_log2fc_2_transname.add(iso)\n",
    "\n",
    "dedf_ak1_iso[dedf_ak1_iso['names'].isin(ak1_nodeg_log2fc_1_yesdex_log2fc_2_transname)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# EdgeR\n",
    "## Aggregate single cells into pseudo-replicates (https://www.sc-best-practices.org/conditions/differential_gene_expression.html#)\n",
    "NUM_OF_CELL_PER_DONOR = 30\n",
    "\n",
    "\n",
    "def aggregate_and_filter(\n",
    "    adata,\n",
    "    cell_identity,\n",
    "    donor_key=\"sample\",\n",
    "    condition_key=\"label\",\n",
    "    cell_identity_key=\"cell_type\",\n",
    "    obs_to_keep=[],  # which additional metadata to keep, e.g. gender, age, etc.\n",
    "    replicates_per_sample=3,\n",
    "):\n",
    "    # subset adata to the given cell identity\n",
    "    adata_cell_pop = adata[adata.obs[cell_identity_key] == cell_identity].copy()\n",
    "    # check which donors to keep according to the number of cells specified with NUM_OF_CELL_PER_DONOR\n",
    "    size_by_donor = adata_cell_pop.obs.groupby([donor_key]).size()\n",
    "    donors_to_drop = [\n",
    "        donor\n",
    "        for donor in size_by_donor.index\n",
    "        if size_by_donor[donor] <= NUM_OF_CELL_PER_DONOR\n",
    "    ]\n",
    "    if len(donors_to_drop) > 0:\n",
    "        print(\"Dropping the following samples:\")\n",
    "        print(donors_to_drop)\n",
    "    df = pd.DataFrame(columns=[*adata_cell_pop.var_names, *obs_to_keep])\n",
    "\n",
    "    adata_cell_pop.obs[donor_key] = adata_cell_pop.obs[donor_key].astype(\"category\")\n",
    "    for i, donor in enumerate(donors := adata_cell_pop.obs[donor_key].cat.categories):\n",
    "        print(f\"\\tProcessing donor {i+1} out of {len(donors)}...\", end=\"\\r\")\n",
    "        if donor not in donors_to_drop:\n",
    "            adata_donor = adata_cell_pop[adata_cell_pop.obs[donor_key] == donor]\n",
    "            # create replicates for each donor\n",
    "            indices = list(adata_donor.obs_names)\n",
    "            random.shuffle(indices)\n",
    "            indices = np.array_split(np.array(indices), replicates_per_sample)\n",
    "            for i, rep_idx in enumerate(indices):\n",
    "                adata_replicate = adata_donor[rep_idx]\n",
    "                # specify how to aggregate: sum gene expression for each gene for each donor and also keep the condition information\n",
    "                agg_dict = {gene: \"sum\" for gene in adata_replicate.var_names}\n",
    "                for obs in obs_to_keep:\n",
    "                    agg_dict[obs] = \"first\"\n",
    "                # create a df with all genes, donor and condition info\n",
    "                df_donor = pd.DataFrame(adata_replicate.X.A)\n",
    "                df_donor.index = adata_replicate.obs_names\n",
    "                df_donor.columns = adata_replicate.var_names\n",
    "                df_donor = df_donor.join(adata_replicate.obs[obs_to_keep])\n",
    "                # aggregate\n",
    "                df_donor = df_donor.groupby(donor_key).agg(agg_dict)\n",
    "                df_donor[donor_key] = donor\n",
    "                df.loc[f\"donor_{donor}_{i}\"] = df_donor.loc[donor]\n",
    "    print(\"\\n\")\n",
    "    # create AnnData object from the df\n",
    "    adata_cell_pop = sc.AnnData(\n",
    "        df[adata_cell_pop.var_names], obs=df.drop(columns=adata_cell_pop.var_names)\n",
    "    )\n",
    "    return adata_cell_pop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "sce <- zellkonverter::readH5AD(\"integrated.h5ad\", verbose=TRUE)\n",
    "#✔ Read ./integrated.h5ad [252ms]\n",
    "#ℹ uns is empty and was skipped\n",
    "#✔ X matrix converted to assay [5.7s]\n",
    "#ℹ layers is empty and was skipped\n",
    "#! The names of these selected var columns have been modified to match R conventions: 'gene_ids-AK1' -> 'gene_ids.AK1', 'gene_ids-NPC' ->\n",
    "#'gene_ids.NPC', and 'gene_ids-iPSC' -> 'gene_ids.iPSC'\n",
    "#✔ var converted to rowData [183ms]\n",
    "#✔ obs converted to colData [48ms]\n",
    "#ℹ varm is empty and was skipped\n",
    "#ℹ obsm is empty and was skipped\n",
    "#ℹ varp is empty and was skipped\n",
    "#ℹ obsp is empty and was skipped\n",
    "#✔ SingleCellExperiment constructed [390ms]\n",
    "#ℹ Skipping conversion of raw\n",
    "#✔ Converting AnnData to SingleCellExperiment ... done\n",
    "#Warning message:\n",
    "#The names of these selected var columns have been modified to match R conventions: 'gene_ids-AK1' -> 'gene_ids.AK1', 'gene_ids-NPC' ->\n",
    "#'gene_ids.NPC', and 'gene_ids-iPSC' -> 'gene_ids.iPSC' \n",
    "\n",
    "sce2seu <- as.Seurat(sce, counts=\"X\", data=NULL)\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(integrated, ['total_counts', 'n_genes_by_counts'], groupby='batch', size=2, log=True, cut=0, inner='quartile', ylabel=['UMI counts', 'Gene counts'], palette=sample_palette, rotation=0.1)\n",
    "sc.pl.violin(integrated, ['pct_counts_mt'], groupby='batch', size=2, log=False, cut=0, inner='quartile', ylabel=['Percentage of mitochondrial genes'], palette=sample_palette, rotation=0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis (Treat them like Bulk-cell Iso-Seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_all = integrated.var[['gene_ids-AK1', 'gene_ids-iPSC', 'gene_ids-NPC', 'total_counts-AK1', 'total_counts-iPSC', 'total_counts-NPC']].copy()\n",
    "#counts_all.to_csv(\"Merged_counts.tab\", sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_all_iso = integrated_iso.var[['gene_ids-AK1', 'gene_ids-iPSC', 'gene_ids-NPC', 'total_counts-AK1', 'total_counts-iPSC', 'total_counts-NPC']].copy()\n",
    "#counts_all_iso.to_csv(\"Merged_counts_iso.tab\", sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cpm_calculator(pdseries):\n",
    "    scale_factor = pdseries.sum()\n",
    "    return pdseries * 10**6 / scale_factor\n",
    "\n",
    "#counts_all = pd.read_table(\"Merged_counts.tab\", index_col=0)\n",
    "for sampleid in ['AK1', 'iPSC', 'NPC']:\n",
    "    counts_all[f'CPM-{sampleid}'] = cpm_calculator(counts_all[f'total_counts-{sampleid}'])\n",
    "\n",
    "\n",
    "#counts_all_iso = pd.read_table(\"Merged_counts_iso.tab\", index_col=0)\n",
    "for sampleid in ['AK1', 'iPSC', 'NPC']:\n",
    "    counts_all_iso[f'CPM-{sampleid}'] = cpm_calculator(counts_all_iso[f'total_counts-{sampleid}'])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unfiltered\n",
    "fig, ax = plt.subplots(1, 1, figsize=(5, 5), constrained_layout=True)\n",
    "sns.violinplot(np.log1p(counts_all.iloc[:, -3:]), palette={'CPM-AK1':'#FF0000', 'CPM-iPSC':'#154360','CPM-NPC':'#229954'}, ax=ax)\n",
    "ax.set_xticklabels(['AK1', 'iPSC', 'NPC'])\n",
    "ax.set_ylabel('log[CPM+1]')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cpm_cutoff = 2\n",
    "counts_all_filtered = counts_all[counts_all.iloc[:, -3:].min(axis=1) >= cpm_cutoff]\n",
    "cpm_all_filtered = counts_all_filtered.iloc[:, -3:].copy()\n",
    "cpm_all_filtered_zero = cpm_all_filtered.fillna(0)\n",
    "cpm_all_filtered_zero_log1p = np.log1p(cpm_all_filtered_zero)\n",
    "cpm_all_filtered_zero_log1p_zscore = cpm_all_filtered_zero_log1p.apply(zscore, axis=1)\n",
    "\n",
    "cpm_cutoff = 2\n",
    "counts_all_iso_filtered = counts_all_iso[counts_all_iso.iloc[:, -3:].min(axis=1) >= cpm_cutoff]\n",
    "cpm_all_iso_filtered = counts_all_iso_filtered.iloc[:, -3:].copy()\n",
    "cpm_all_iso_filtered_zero = cpm_all_iso_filtered.fillna(0)\n",
    "cpm_all_iso_filtered_zero_log1p = np.log1p(cpm_all_iso_filtered_zero)\n",
    "cpm_all_iso_filtered_zero_log1p_zscore = cpm_all_iso_filtered_zero_log1p.apply(zscore, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maybe centering is needed here\n",
    "clustergrid = sns.clustermap(cpm_all_filtered_zero_log1p_zscore, method='ward', metric='euclidean', cmap=cmap)\n",
    "clustergrid.ax_heatmap.set_xticklabels(['AK1', 'iPSC', 'NPC'])\n",
    "\n",
    "clustergrid = sns.clustermap(cpm_all_iso_filtered_zero_log1p_zscore, method='ward', metric='euclidean', cmap=cmap)\n",
    "clustergrid.ax_heatmap.set_xticklabels(['AK1', 'iPSC', 'NPC'])\n",
    "\n",
    "\n",
    "\n",
    "# No DEG (|log2FC|<1) Yes DEX (|log2FC|>2)\n",
    "# Isoforms (zscore)\n",
    "clustergrid = sns.clustermap(cpm_all_iso_filtered_zero_log1p_zscore[cpm_all_iso_filtered_zero_log1p_zscore.index.isin(ak1_nodeg_log2fc_1_yesdex_log2fc_2_transname)], method='ward', metric='euclidean', cmap='viridis', xticklabels=True, yticklabels=False)\n",
    "clustergrid.ax_heatmap.set_xticklabels(['AK1', 'iPSC', 'NPC'])\n",
    "# Isoforms (just log1p)\n",
    "clustergrid = sns.clustermap(cpm_all_iso_filtered_zero_log1p[cpm_all_iso_filtered_zero_log1p.index.isin(ak1_nodeg_log2fc_1_yesdex_log2fc_2_transname)], method='ward', metric='euclidean', cmap='viridis', xticklabels=True, yticklabels=False)\n",
    "clustergrid.ax_heatmap.set_xticklabels(['AK1', 'iPSC', 'NPC'])\n",
    "# Genes (zscore)=> it lloks like they are DEG...\n",
    "clustergrid = sns.clustermap(cpm_all_filtered_zero_log1p_zscore[cpm_all_filtered_zero_log1p_zscore.index.isin(ak1_nodeg_log2fc_1_yesdex_log2fc_2_genename)], method='ward', metric='euclidean', cmap='viridis', xticklabels=True, yticklabels=False)\n",
    "clustergrid.ax_heatmap.set_xticklabels(['AK1', 'iPSC', 'NPC'])\n",
    "# Genes (just log1p)=> it lloks like they are DEG...\n",
    "clustergrid = sns.clustermap(cpm_all_filtered_zero_log1p[cpm_all_filtered_zero_log1p.index.isin(ak1_nodeg_log2fc_1_yesdex_log2fc_2_genename)], method='ward', metric='euclidean', cmap='viridis', xticklabels=True, yticklabels=False)\n",
    "clustergrid.ax_heatmap.set_xticklabels(['AK1', 'iPSC', 'NPC'])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### First, execute *_Run_DropletUtils.R*"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing on filtered CBC-UMI matrix"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Doublet detection (using scrublet version 0.2.3) and removal for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample in [ak1, ipsc, npc]:\n",
    "    sce.pp.scrublet(sample, adata_sim=None, sim_doublet_ratio=2.0, expected_doublet_rate=0.08, stdev_doublet_rate=0.02, synthetic_doublet_umi_subsampling=1.0, knn_dist_metric='euclidean', n_prin_comps=30, verbose=True)\n",
    "\n",
    "'''\n",
    "# Filtering\n",
    "ak1 = ak1[ak1.obs['predicted_doublet'] == False, :]\n",
    "ipsc = ipsc[ipsc.obs['predicted_doublet'] == False, :]\n",
    "npc = npc[npc.obs['predicted_doublet'] == False, :]\n",
    "\n",
    "for sample in [ak1, ipsc, npc]:\n",
    "    sample.obs['n_counts'] = sample.X.sum(axis=1)\n",
    "    sample.obs['n_genes'] = (sample.X > 0).sum(axis=1)\n",
    "\n",
    "q20 = ak1.obs['n_counts'].quantile(q=0.20, interpolation='linear') # 20% quantile value\n",
    "ak1 = ak1[ak1.obs['n_counts'] > q20, :]\n",
    "\n",
    "q20 = ipsc.obs['n_counts'].quantile(q=0.20, interpolation='linear') # 20% quantile value\n",
    "ipsc = ipsc[ipsc.obs['n_counts'] > q20, :]\n",
    "\n",
    "q20 = npc.obs['n_counts'].quantile(q=0.20, interpolation='linear') # 20% quantile value\n",
    "npc = npc[npc.obs['n_counts'] > q20, :]\n",
    "'''"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Batch integration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "integrated = AnnData.concatenate(ak1, ipsc, npc, join='outer', batch_categories = ['AK1', 'iPSC', 'NPC'], index_unique = '-')\n",
    "integrated.raw = integrated\n",
    "integrated.var['mt'] = integrated.var_names.str.startswith('MT-')\n",
    "sc.pp.calculate_qc_metrics(integrated, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "integrated.var[integrated.var.index.isin(integrated.var_names[integrated.var_names.str.contains('\\+')])]\n",
    "#Checked Distribution of UMI counts and Gene counts\n",
    "#integrated.obs['n_counts'] = integrated.X.sum(1)\n",
    "#integrated.obs['n_genes'] = (integrated.X > 0).sum(1)\n",
    "#sns.set(font=\"Arial\", font_scale=1.5, style='ticks')\n",
    "#sc.pl.violin(integrated, ['n_counts', 'n_genes'], groupby='batch', size=2, log=True, cut=0, inner='quartile', ylabel=['UMI counts', 'Gene counts'], rotation=0.1)\n",
    "#sns.despine()\n",
    "#sns.set(font=\"Arial\", font_scale=1, style='ticks') # Back to original settings\n",
    "\n",
    "# For making background gene-set for performing GSEA\n",
    "with open(\"Background_GeneSets_from_MAS-ISO-Seq.txt\", 'w') as rfh:\n",
    "    for i in integrated.var.index:\n",
    "        rfh.write(i + '\\n')\n",
    "        rfh.flush()\n",
    "\n",
    "sc.pl.violin(integrated, ['total_counts', 'n_genes_by_counts'], groupby='batch', size=2, log=True, cut=0, inner='quartile', ylabel=['UMI counts', 'Gene counts'], palette=sample_palette, rotation=0.1)\n",
    "sc.pl.violin(integrated, ['pct_counts_mt'], groupby='batch', size=2, log=False, cut=0, inner='quartile', ylabel=['Percentage of mitochondrial genes'], palette=sample_palette, rotation=0.1)\n",
    "\n",
    "cpm_counts = sc.pp.normalize_total(integrated, target_sum=1e6, inplace=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_genes(integrated, min_cells=0) # 'n_cells' added in integrated.var \n",
    "sc.pp.filter_genes(integrated, min_cells=5) # 'n_cells' added in integrated.var \n",
    "\n",
    "cpm_counts = sc.pp.normalize_total(integrated, target_sum=1e6, inplace=False)\n",
    "\n",
    "'''\n",
    "In [142]: cpm_counts\n",
    "Out[142]: \n",
    "{'X': <24634x45853 sparse matrix of type '<class 'numpy.float32'>'\n",
    " \twith 17515606 stored elements in Compressed Sparse Row format>,\n",
    " 'norm_factor': array([2330., 1335., 3835., ...,  200.,  283.,  204.], dtype=float32)}\n",
    " \n",
    "cpm_counts['norm_factor'] == integrated.obs['total_counts'] # CPM = counts per million (gene count*10^6/total counts)\n",
    "'''\n",
    "integrated.layers[\"cpm\"] = cpm_counts[\"X\"]\n",
    "integrated.layers[\"log1p_cpm\"] = sc.pp.log1p(cpm_counts[\"X\"], copy=True)\n",
    "\n",
    "fig, axes = plt.subplots(1, 2, figsize=(8, 4), constrained_layout=True)\n",
    "p1 = sns.histplot(integrated.obs[\"total_counts\"], bins=100, kde=True, ax=axes[0])\n",
    "axes[0].set_title(\"Total counts\")\n",
    "axes[0].set_xlabel('Total counts in each cell')\n",
    "axes[0].xaxis.set_major_formatter(StrMethodFormatter('{x:,.0f}'))\n",
    "axes[0].yaxis.set_major_formatter(StrMethodFormatter('{x:,.0f}'))\n",
    "p2 = sns.histplot(integrated.layers[\"log1p_cpm\"].sum(1), bins=100, kde=True, ax=axes[1])\n",
    "axes[1].set_title(\"Shifted logarithm (CPM)\")\n",
    "axes[1].set_xlabel('Total log(CPM+1) in each cell')\n",
    "axes[1].xaxis.set_major_formatter(StrMethodFormatter('{x:,.0f}'))\n",
    "axes[1].yaxis.set_major_formatter(StrMethodFormatter('{x:,.0f}'))\n",
    "axes[1].legend().remove()\n",
    "'''p3 = sns.histplot(integrated.layers[\"cpm\"].sum(1), bins=100, kde=True, ax=axes[2])\n",
    "axes[2].set_title(\"CPM\")\n",
    "axes[2].legend().remove()\n",
    "This would just be collection of 10^6\n",
    "'''\n",
    "\n",
    "# By Sample\n",
    "fig, axes = plt.subplots(1, 3, figsize=(12, 4), constrained_layout=True)\n",
    "p1 = sns.histplot(integrated[integrated.obs['batch'] == 'AK1'].layers[\"log1p_cpm\"].sum(1), bins=100, kde=True, ax=axes[0])\n",
    "axes[0].set_title(\"AK1\")\n",
    "axes[0].set_xlabel('Total log(CPM+1) in each cell')\n",
    "axes[0].xaxis.set_major_formatter(StrMethodFormatter('{x:,.0f}'))\n",
    "axes[0].yaxis.set_major_formatter(StrMethodFormatter('{x:,.0f}'))\n",
    "axes[0].legend().remove()\n",
    "p2 = sns.histplot(integrated[integrated.obs['batch'] == 'iPSC'].layers[\"log1p_cpm\"].sum(1), bins=100, kde=True, ax=axes[1])\n",
    "axes[1].set_title(\"iPSC\")\n",
    "axes[1].set_xlabel('Total log(CPM+1) in each cell')\n",
    "axes[1].xaxis.set_major_formatter(StrMethodFormatter('{x:,.0f}'))\n",
    "axes[1].yaxis.set_major_formatter(StrMethodFormatter('{x:,.0f}'))\n",
    "axes[1].legend().remove()\n",
    "p3 = sns.histplot(integrated[integrated.obs['batch'] == 'NPC'].layers[\"log1p_cpm\"].sum(1), bins=100, kde=True, ax=axes[2])\n",
    "axes[2].set_title(\"NPC\")\n",
    "axes[2].set_xlabel('Total log(CPM+1) in each cell')\n",
    "axes[2].xaxis.set_major_formatter(StrMethodFormatter('{x:,.0f}'))\n",
    "axes[2].yaxis.set_major_formatter(StrMethodFormatter('{x:,.0f}'))\n",
    "axes[2].legend().remove()\n",
    "#integrated.write(filename='/data/Projects/phenomata/01.Projects/13.AK1_PacBio/02.RNA/integrated_20230304.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "integrated.X = integrated.layers['log1p_norm'].copy()\n",
    "sc.pp.highly_variable_genes(integrated)\n",
    "integrated.var['highly_variable'].value_counts() # 3,494 ==> 2023-10-13\n",
    "\n",
    "integrated.raw = integrated\n",
    "#sc.pp.scale(integrated, max_value=10) # tabula muris senis default (2021-08-10) # mean and std on adata.var\n",
    "#sc.pp.scale(test3, zero_center=True, max_value=10, copy=False, layer=None, obsm=None)\n",
    "\n",
    "np.random.seed(0)\n",
    "rand_is = np.random.permutation(list(range(integrated.shape[0])))\n",
    "\n",
    "sc.tl.pca(integrated, n_comps=100, use_highly_variable=True, svd_solver='arpack')\n",
    "sc.pl.pca(integrated, color=['batch', 'total_counts'], legend_loc='on data', size=15, add_outline=False, color_map='CMRmap', palette=sample_palette, components=['1,2'], title=['', 'UMI count'], annotate_var_explained=True)\n",
    "sc.pl.pca(integrated, color=['batch', 'total_counts'], legend_loc='on data', size=15, add_outline=False, color_map='CMRmap', palette=sample_palette, components=['1,3'], title=['', 'UMI count'], annotate_var_explained=True)\n",
    "sc.pl.pca(integrated, color=['batch', 'total_counts'], legend_loc='on data', size=15, add_outline=False, color_map='CMRmap', palette=sample_palette, components=['1,4'], title=['', 'UMI count'], annotate_var_explained=True)\n",
    "sc.pl.pca_variance_ratio(integrated, n_pcs=100, log=True)\n",
    "sc.pl.pca_loadings(integrated, components='1,2,3,4', include_lowest=True)\n",
    "\n",
    "cell_cycle_genes=[x.strip()[0] + x.strip()[1:] for x in open(\"/data/Projects/phenomata/01.Projects/11.Vascular_Aging/Database/regev_lab_cell_cycle_genes.txt\")]\n",
    "s_genes= cell_cycle_genes[:43]\n",
    "g2m_genes= cell_cycle_genes[43:]\n",
    "cell_cycle_genes = [x for x in cell_cycle_genes if x in integrated.var_names]\n",
    "sc.tl.score_genes_cell_cycle(integrated, s_genes=s_genes, g2m_genes=g2m_genes)\n",
    "\"\"\"\n",
    "Used 'raw' attribute of adata (use_raw = True if .raw is present)\n",
    "So, log-tranformed scran-normalized counts are put into score_genes_cell_cycle function\n",
    "\"\"\"\n",
    "\n",
    "df = integrated.obs[['batch', 'phase']]\n",
    "ax = pd.crosstab(df['batch'], df['phase'], normalize='index', margins=True).plot.bar(stacked=True, rot=45, color={'S': '#689aff', 'G2M': '#fdbf6f', 'G1': '#b15928'})\n",
    "ax.legend(loc='upper left', bbox_to_anchor=(1.01, 0.25), frameon=False)\n",
    "ax.set_ylabel('Proportion of Cell Cycle Phase')\n",
    "ax.set_xlabel(\"\")\n",
    "plt.tight_layout()\n",
    "sns.despine()\n",
    "\n",
    "sc.pp.neighbors(integrated, n_neighbors=15, n_pcs=10)\n",
    "sc.tl.umap(integrated)\n",
    "sc.pl.umap(integrated, color=['batch', 'total_counts'], legend_loc='on data', size=15, add_outline=False, color_map='CMRmap', palette=sample_palette, components=['1,2'], title=['', 'UMI count'])\n",
    "\n",
    "sc.tl.draw_graph(integrated, layout='fa', init_pos=None, neighbors_key=None) ## init_pos가 .obsm에 있는 pca, umap, paga 등이 될 수 있다.\n",
    "sc.pl.draw_graph(integrated, color=['batch', 'total_counts'], add_outline=True, legend_loc='on data', size=15, color_map='CMRmap')\n",
    "\n",
    "'''\n",
    "#sce.pp.bbknn default ==> n_pcs=50, neighbors_within_batch=3, trim=None, annoy_n_trees=10,\n",
    "sce.pp.bbknn(integrated, batch_key='batch', n_pcs=20, neighbors_within_batch=5, trim=None)\n",
    "sc.tl.umap(integrated, min_dist=0.5, spread=1.0, n_components=2, alpha=1.0, gamma=1.0, init_pos='spectral', method='umap')\n",
    "#integrated.uns['batch_colors'] = ['#689aff', '#fdbf6f', '#b15928']\n",
    "sc.pl.umap(integrated, color=['batch'], add_outline=False, legend_loc='on data', size=20, title='')\n",
    "sns.despine()\n",
    "'''\n",
    "fig, axes = plt.subplots(3,1,figsize=(4.5,15.5))\n",
    "for i in range(len(integrated.obs['batch'].cat.categories)):\n",
    "    sc.pl.umap(integrated, color=['batch'], add_outline=False, legend_loc=None, groups=integrated.obs['batch'].cat.categories[i], title=integrated.obs['batch'].cat.categories[i], size=20, ax=axes[i])\n",
    "    sns.despine(ax=axes[i])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(integrated, 'batch', method='wilcoxon', corr_method='benjamini-hochberg', use_raw=True, pts=True, key_added='DEG_bw_batches_wilcoxon') # key_added=''\n",
    "\n",
    "sc.pl.rank_genes_groups(integrated, n_genes=15, sharey=False, key='DEG_bw_batches_wilcoxon')\n",
    "\n",
    "ax_dict = sc.pl.rank_genes_groups_heatmap(integrated, n_genes=15, groupby='batch', key='DEG_bw_batches_wilcoxon', groups=['AK1', 'iPSC', 'NPC'], show_gene_labels=True, min_logfoldchange=1, dendrogram=False, cmap='viridis', use_raw=False, swap_axes=True, show=False, var_group_rotation=90)\n",
    "ax_dict['heatmap_ax'].set_yticklabels(labels=ax_dict['heatmap_ax'].get_yticklabels(), fontstyle='italic')\n",
    "\n",
    "\n",
    "markers = [\"Pecam1\", \"Cdh5\", \"Nos3\", \"Acta2\", \"Cnn1\", \"Tagln\", \"Rgs5\", \"Kcnj8\", \"Col1a1\", \"Col5a1\", \"Dpt\", \"Cd19\", \"Ighm\", \"Cd14\", \"Cd68\", \"Cd3d\"] # Cd3g 없음\n",
    "sc.pl.stacked_violin(test3, markers, groupby='batch')\n",
    "\n",
    "result = test3.uns['rank_genes_groups']\n",
    "groups = result['names'].dtype.names\n",
    "deg_wilcoxon = pd.DataFrame({group + '_' + key: result[key][group] for group in groups for key in ['names', 'logfoldchanges', 'scores', 'pvals_adj']})\n",
    "deg_wilcoxon.to_csv(\"/data/Projects/phenomata/01.Projects/11.Vascular_Aging/03.Scanpy/20210916_scanpy_deg.csv\", mode='w')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# AK1 (B-lymphocyte) specific gene\n",
    "genes = ['CD19', 'MS4A1'] # MS4A1 (CD20)\n",
    "fig, axes = plt.subplots(1, 2, figsize=(10, 5), constrained_layout=True)\n",
    "for i in range(len(genes)):\n",
    "    sc.pl.umap(integrated, color=genes[i], add_outline=False, legend_loc='right margin', size=40, color_map='viridis', use_raw=True, ax=axes[i])\n",
    "    axes[i].set_title(genes[i], style='italic')\n",
    "    sns.despine(ax=axes[i])\n",
    "\n",
    "# iPSC-specific gene\n",
    "genes = ['POU5F1', 'SOX2', 'NANOG'] # JARID2\n",
    "fig, axes = plt.subplots(1, 3, figsize=(15, 10), constrained_layout=True)\n",
    "for i in range(len(genes)):\n",
    "    sc.pl.umap(integrated, color=genes[i], add_outline=False, legend_loc='right margin', size=40, color_map='viridis', use_raw=True, ax=axes[i])\n",
    "    axes[i].set_title(genes[i], style='italic')\n",
    "    sns.despine(ax=axes[i])\n",
    "\n",
    "\n",
    "# NPC-specific gene\n",
    "genes = ['NCAM1', 'KMT2D']\n",
    "fig, axes = plt.subplots(1, 2, figsize=(10, 5), constrained_layout=True)\n",
    "for i in range(len(genes)):\n",
    "    sc.pl.umap(integrated, color=genes[i], add_outline=False, legend_loc='right margin', size=40, color_map='viridis', use_raw=True, ax=axes[i])\n",
    "    axes[i].set_title(genes[i], style='italic')\n",
    "    sns.despine(ax=axes[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.tl.leiden(test3, resolution=0.5, key_added='leiden_r05') #### 0 ~ 13 ==> 2021-09-28\n",
    "sc.tl.leiden(test3, resolution=1.0, key_added='leiden_r10')\n",
    "sc.pl.umap(test3, color=['batch', 'leiden_r05', 'leiden_r10'], add_outline=False, legend_loc='right margin', size=20)\n",
    "\n",
    "fig, axes = plt.subplots(1,3)\n",
    "sc.pl.umap(test3, color=['batch'], add_outline=False, legend_loc='right margin', size=20, groups=['m01'], title='1 month', ax=axes[0])\n",
    "sc.pl.umap(test3, color=['batch'], add_outline=False, legend_loc='right margin', size=20, groups=['m10'], title='10 months', ax=axes[1])\n",
    "sc.pl.umap(test3, color=['batch'], add_outline=False, legend_loc='right margin', size=20, groups=['m20'], title='20 months', ax=axes[2])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Diffusion pseudotime\n",
    "sc.tl.diffmap(test3_endo)\n",
    "sc.pl.diffmap(test3_endo, color=['batch', 'Pecam1', 'Cdh5'], add_outline=False, legend_loc='right margin', size=70, color_map='CMRmap')\n",
    "\n",
    "sc.tl.draw_graph(test3, layout='fa', init_pos=None, neighbors_key=None) ## init_pos가 .obsm에 있는 pca, umap, paga 등이 될 수 있다.\n",
    "sc.pl.draw_graph(test3, color=['batch', 'PECAM1', 'CDH5', 'phase'], add_outline=True, legend_loc='right margin', size=10, color_map='CMRmap')\n",
    "\n",
    "start_cell = np.isin(test3_endo.obs['endo_leiden_r05'], '0') # boolean numpy array ==> array([False, False, False, ..., False, False, False])\n",
    "#max_start_id = np.argmin(test3_endo.obsm['X_diffmap'][start_cell,1]) # 262\n",
    "max_start_id = np.argmax(test3_endo.obsm['X_diffmap'][start_cell,1])\n",
    "root_id = np.arange(len(start_cell))[start_cell][max_start_id] # 341\n",
    "test3_endo.uns['iroot'] = root_id\n",
    "\n",
    "sc.tl.dpt(test3_endo, n_branchings=1, n_dcs=10) # n_branchings를 0으로 하면 (recommended by Scanpy developer) dpt_groups가 생성 안 됨.\n",
    "#computing Diffusion Pseudotime using n_dcs=10\n",
    "sc.pl.dpt_groups_pseudotime(test3_endo) # 여기에서 pseudotime trajecgory 확인.\n",
    "\n",
    "lin = ('2', '0', '3', '1') # DPT pseudotime group ordering에 맞게 배치\n",
    "test3_endo.obs['dpt_groups'] = test3_endo.obs['dpt_groups'].cat.reorder_categories(list(lin), ordered=True)\n",
    "sc.pl.dpt_groups_pseudotime(test3_endo) # 다시 ordering에 맞게 plotting\n",
    "sc.pl.dpt_timeseries(test3_endo[:, test3_endo.var.highly_variable])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Diffusion pseudotime (Experimental)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.diffmap(integrated)\n",
    "sc.pl.diffmap(integrated, color=['batch'], add_outline=False, legend_loc='right margin', size=70, color_map='CMRmap')\n",
    "\n",
    "\n",
    "sc.tl.draw_graph(test3, layout='fa', init_pos=None, neighbors_key=None) ## init_pos가 .obsm에 있는 pca, umap, paga 등이 될 수 있다.\n",
    "sc.pl.draw_graph(test3, color=['batch', 'PECAM1', 'CDH5', 'phase'], add_outline=True, legend_loc='right margin', size=10, color_map='CMRmap')\n",
    "\n",
    "start_cell = np.isin(test3_endo.obs['endo_leiden_r05'], '0') # boolean numpy array ==> array([False, False, False, ..., False, False, False])\n",
    "#max_start_id = np.argmin(test3_endo.obsm['X_diffmap'][start_cell,1]) # 262\n",
    "max_start_id = np.argmax(test3_endo.obsm['X_diffmap'][start_cell,1])\n",
    "root_id = np.arange(len(start_cell))[start_cell][max_start_id] # 341\n",
    "test3_endo.uns['iroot'] = root_id\n",
    "\n",
    "sc.tl.dpt(test3_endo, n_branchings=1, n_dcs=10) # n_branchings를 0으로 하면 (recommended by Scanpy developer) dpt_groups가 생성 안 됨.\n",
    "#computing Diffusion Pseudotime using n_dcs=10\n",
    "sc.pl.dpt_groups_pseudotime(test3_endo) # 여기에서 pseudotime trajecgory 확인.\n",
    "\n",
    "lin = ('2', '0', '3', '1') # DPT pseudotime group ordering에 맞게 배치\n",
    "test3_endo.obs['dpt_groups'] = test3_endo.obs['dpt_groups'].cat.reorder_categories(list(lin), ordered=True)\n",
    "sc.pl.dpt_groups_pseudotime(test3_endo) # 다시 ordering에 맞게 plotting\n",
    "sc.pl.dpt_timeseries(test3_endo[:, test3_endo.var.highly_variable])\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Force-Directged Graph (Experimental)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.draw_graph(integrated, layout='fa', init_pos=None, neighbors_key=None) ## init_pos가 .obsm에 있는 pca, umap, paga 등이 될 수 있다.\n",
    "sc.pl.draw_graph(integrated, color=['batch'], add_outline=True, legend_loc='on data', size=40, color_map='CMRmap')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "python",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
