{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "cmap = matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#FFFF00\", \"#000000\", \"#0066CC\"])\n",
    "\n",
    "sns.set_theme(font=\"Arial\", font_scale=1.15, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n",
    "#%matplotlib inline\n",
    "sample_palette = {'AK1':'#FF0000', 'iPSC':'#154360', 'H1': '#40E0D0', 'HUES64': '#437299', 'NPC':'#229954', 'HG002': '#545454'}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#dir=\"/Users/mhryansohn/Desktop/01.Workspace/01.Projects/03.AK1-PacBio/01.DNA/Merged_Run/DNA_methylation/PMD_from_AK1_WGBS/\"\n",
    "dir=\"/mnt/data/Projects/phenomata/01.Projects/13.AK1_PacBio/01.DNA/Merged/Phasing_wIndels/MethPipe/PMD/\"\n",
    "ID, Length = list(), list()\n",
    "with open(f\"{dir}AK1_PMD_HiFi.bed\", 'r') as pmdf:\n",
    "    for line in pmdf:  \n",
    "        line = line.strip('\\n').split('\\t')\n",
    "    \n",
    "        id = f\"{line[0]}:{line[1]}-{line[2]}_AK1\"\n",
    "        length = int(line[2]) - int(line[1])\n",
    "    \n",
    "        ID.append(id)\n",
    "        Length.append(length)\n",
    "\n",
    "pmd_from_ak1_hifi = pd.DataFrame(list(zip(ID, Length)), columns=['ID', 'Length']).set_index('ID')\n",
    "del ID, Length\n",
    "\n",
    "ID, Length = list(), list()\n",
    "with open(f\"{dir}HG002_PMD_HiFi.bed\", 'r') as pmdf:\n",
    "    for line in pmdf:  \n",
    "        line = line.strip('\\n').split('\\t')\n",
    "    \n",
    "        id = f\"{line[0]}:{line[1]}-{line[2]}_HG002\"\n",
    "        length = int(line[2]) - int(line[1])\n",
    "    \n",
    "        ID.append(id)\n",
    "        Length.append(length)\n",
    "\n",
    "pmd_from_hg002_hifi = pd.DataFrame(list(zip(ID, Length)), columns=['ID', 'Length']).set_index('ID')\n",
    "del ID, Length\n",
    "\n",
    "pmd_total = pd.concat([pmd_from_ak1_hifi, pmd_from_hg002_hifi], axis=0)\n",
    "pmd_total = pmd_total.assign(Sample=pmd_total.index.str.split('_').str[-1])\n",
    "\n",
    "'''\n",
    "fig, ax = plt.subplots(figsize=(8,8))\n",
    "sns.histplot(data=pmd_from_ak1_wgbs, x='Length', kde=True, stat='count', ax=ax)\n",
    "ax.set_xlim(pmd_from_ak1_wgbs['Length'].min(), pmd_from_ak1_wgbs['Length'].max())\n",
    "ax.set_xlabel(\"Length of PMD (bp)\")\n",
    "sns.despine(ax=ax)\n",
    "'''\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8,8), constrained_layout=True)\n",
    "hist = sns.histplot(data=pmd_total, x='Length', hue='Sample', kde=True, stat='count', palette=sample_palette, ax=ax)\n",
    "ax.set_xlim(pmd_total['Length'].min(), pmd_total['Length'].max())\n",
    "ax.set_xlabel(f\"Length of PMD\")\n",
    "ax.set_ylabel(\"The Number of PMD\")\n",
    "legend = ax.legend_\n",
    "handles = legend.legendHandles\n",
    "labels = [text.get_text() for text in legend.texts]\n",
    "\n",
    "ax.legend(handles, labels, loc='upper right', bbox_to_anchor=(0.9, 0.8), frameon=True, fancybox=False, edgecolor='black', prop={'size':15}, title=None)\n",
    "plt.savefig('PMD_Length_Distribution_AK1_HG002.png')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PMD (from AK1 HiFi) mCG Merging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir='/mnt/data/Projects/phenomata/01.Projects/13.AK1_PacBio/01.DNA/Merged/Phasing_wIndels/MethPipe/PMD/'\n",
    "\n",
    "ak1 = pd.read_table(f\"{dir}AK1_AK1_HiFi_PMD_met.txt\", index_col=0)\n",
    "#hg002 = pd.read_table(f\"{dir}HG002_merged_AK1_WGBS_PMD_met.txt\", index_col=0)\n",
    "#hg002 = hg002.rename(columns={'HG002_merged': 'HG002'})\n",
    "hg002 = pd.read_table(f\"{dir}HG002_AK1_HiFi_PMD_met.txt\", index_col=0)\n",
    "ipsc = pd.read_table(f\"{dir}iPSC_AK1_HiFi_PMD_met.txt\", index_col=0)\n",
    "h1 = pd.read_table(f\"{dir}H1_AK1_HiFi_PMD_met.txt\", index_col=0)\n",
    "hues64 = pd.read_table(f\"{dir}HUES64_AK1_HiFi_PMD_met.txt\", index_col=0)\n",
    "npc = pd.read_table(f\"{dir}NPC_AK1_HiFi_PMD_met.txt\", index_col=0)\n",
    "merge = pd.concat([ak1, hg002, ipsc, h1, hues64, npc], axis=1)\n",
    "\n",
    "merge = merge[~merge.index.str.startswith('chrY')]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main Figure\n",
    "\n",
    "sns.set_theme(font=\"Arial\", font_scale=1.5, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10,15), constrained_layout=True)\n",
    "heatmap = sns.heatmap(merge.iloc[:, [0, 2, 3, 4, 5]], \n",
    "                      vmin=0, \n",
    "                      vmax=100, \n",
    "                      xticklabels=True, \n",
    "                      yticklabels=False, \n",
    "                      cbar=True, \n",
    "                      cbar_kws={'label':'DNA methylation (%)'}, \n",
    "                      cmap=cmap, ax=ax)\n",
    "heatmap.collections[0].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=20)\n",
    "ax.set_ylabel('PMDs from AK1 HiFi (N=2,087)', labelpad=15)\n",
    "for separator in [1, 2, 4]:\n",
    "    heatmap.add_patch(plt.Rectangle((separator, 0), 0.05, merge.shape[0], fill=True, color=\"white\", lw=0))\n",
    "    \n",
    "plt.savefig('PMD_AK1_HiFi_mCG_Heatmap.png')\n",
    "\n",
    "# Main Figure (w/ HG002)\n",
    "fig, ax = plt.subplots(figsize=(10,15))\n",
    "heatmap = sns.heatmap(merge, \n",
    "                      vmin=0, \n",
    "                      vmax=100, \n",
    "                      xticklabels=True, \n",
    "                      yticklabels=False, \n",
    "                      cbar=True, \n",
    "                      cbar_kws={'label':'DNA methylation (%)'}, \n",
    "                      cmap=cmap, ax=ax)\n",
    "heatmap.collections[0].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=20)\n",
    "ax.set_ylabel('PMDs from AK1 HiFi (N=2,087)', labelpad=15)\n",
    "plt.savefig('PMD_AK1_HiFi_mCG_Heatmap_wHG002.png')\n",
    "\n",
    "sns.set(font=\"Arial\", font_scale=1.15, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 100kb DNA methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "sns.set(font=\"Arial\", font_scale=2, style='ticks')\n",
    "\n",
    "#df1 = pd.read_table('100kb_individual/AK1_100kbmet.tab', index_col=0, header=None)\n",
    "#df2 = pd.read_table('100kb_individual/iPSC_100kbmet.tab', index_col=0, header=None)\n",
    "#df3 = pd.read_table('100kb_individual/H1_100kbmet.tab', index_col=0, header=None)\n",
    "#df4 = pd.read_table('100kb_individual/NPC_100kbmet.tab', index_col=0, header=None)\n",
    "#\n",
    "\n",
    "#df = pd.concat([df1, df2, df3, df4], axis=1)\n",
    "\n",
    "# Main Figure (Fig.1 Violinplot)\n",
    "ak1_100kb = pd.read_table('AK1_100kbmet.tab', index_col=0, header=None)\n",
    "hg002_100kb = pd.read_table('HG002_100kbmet.tab', index_col=0, header=None) # Added For Fig.S\n",
    "ipsc_100kb = pd.read_table('iPSC_100kbmet.tab', index_col=0, header=None)\n",
    "h1_100kb = pd.read_table('H1_100kbmet.tab', index_col=0, header=None)\n",
    "hues64_100kb = pd.read_table('HUES64_100kbmet.tab', index_col=0, header=None) # Added 20230703\n",
    "npc_100kb = pd.read_table('NPC_100kbmet.tab', index_col=0, header=None)\n",
    "\n",
    "df = pd.concat([ak1_100kb, ipsc_100kb, h1_100kb, hues64_100kb, npc_100kb], axis=1)\n",
    "df.columns = ['AK1', 'iPSC', 'H1', 'HUES64', 'NPC']\n",
    "df.index.name = 'Region'\n",
    "df = df.dropna()\n",
    "\n",
    "fig, axes = plt.subplots(1, 1, figsize=(8, 5), constrained_layout=True)\n",
    "sns.violinplot(data=df, palette=sample_palette, ax=axes)\n",
    "axes.set_ylabel('100kb mCG Level (%)')\n",
    "sns.despine(ax=axes)\n",
    "#plt.savefig('Violin_100kb_met.png')\n",
    "\n",
    "# Suppl Figure (Fig.SX Violinplot)\n",
    "df = pd.concat([ak1_100kb, hg002_100kb, ipsc_100kb, h1_100kb, hues64_100kb, npc_100kb], axis=1)\n",
    "df.columns = ['AK1', 'HG002', 'iPSC', 'H1', 'HUES64', 'NPC']\n",
    "df.index.name = 'Region'\n",
    "df = df.dropna()\n",
    "\n",
    "fig, axes = plt.subplots(1, 1, figsize=(8, 5), constrained_layout=True)\n",
    "sns.violinplot(data=df, palette=sample_palette, ax=axes)\n",
    "axes.set_ylabel('100kb mCG Level (%)')\n",
    "sns.despine(ax=axes)\n",
    "#plt.savefig('Violin_100kb_met_wHG002.png')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10kb DNA methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sns.set(font=\"Arial\", font_scale=2, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n",
    "\n",
    "sample_palette = {'AK1':'#FF0000', 'iPSC':'#154360', 'H1': '#40E0D0', 'HUES64': '#437299', 'NPC':'#229954', 'HG002': '#545454'}\n",
    "\n",
    "# Main Figure (Fig.1 Violinplot)\n",
    "ak1_10kb = pd.read_table('AK1_10kbmet.tab', index_col=0, header=None)\n",
    "hg002_10kb = pd.read_table('HG002_10kbmet.tab', index_col=0, header=None) # Added For Fig.S\n",
    "ipsc_10kb = pd.read_table('iPSC_10kbmet.tab', index_col=0, header=None)\n",
    "h1_10kb = pd.read_table('H1_10kbmet.tab', index_col=0, header=None)\n",
    "hues64_10kb = pd.read_table('HUES64_10kbmet.tab', index_col=0, header=None)\n",
    "npc_10kb = pd.read_table('NPC_10kbmet.tab', index_col=0, header=None)\n",
    "\n",
    "df = pd.concat([ak1_10kb, ipsc_10kb, h1_10kb, hues64_10kb, npc_10kb], axis=1)\n",
    "df.columns = ['AK1', 'iPSC', 'H1', 'HUES64', 'NPC']\n",
    "df.index.name = 'Region'\n",
    "df = df.dropna()\n",
    "\n",
    "fig, axes = plt.subplots(1, 1, figsize=(8, 5), constrained_layout=True)\n",
    "sns.violinplot(data=df, palette=sample_palette, cut=0, ax=axes)\n",
    "axes.set_ylabel('10kb mCG Level (%)')\n",
    "plt.savefig('Violin_10kb_met.png')\n",
    "\n",
    "# Suppl Figure (Fig.SX Violinplot)\n",
    "df = pd.concat([ak1_10kb, hg002_10kb, ipsc_10kb, h1_10kb, hues64_10kb, npc_10kb], axis=1)\n",
    "df.columns = ['AK1', 'HG002', 'iPSC', 'H1', 'HUES64', 'NPC']\n",
    "df.index.name = 'Region'\n",
    "df = df.dropna()\n",
    "\n",
    "fig, axes = plt.subplots(1, 1, figsize=(8, 5), constrained_layout=True)\n",
    "sns.violinplot(data=df, palette=sample_palette, cut=0, ax=axes)\n",
    "axes.set_ylabel('10kb mCG Level (%)')\n",
    "plt.savefig('Violin_10kb_met_wHG002.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA using 10kb methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_10kb = pd.concat([ak1_10kb, hg002_10kb, ipsc_10kb, h1_10kb, hues64_10kb, npc_10kb], axis=1)\n",
    "df_10kb.columns = ['AK1', 'HG002', 'iPSC', 'H1', 'HUES64', 'NPC']\n",
    "df_10kb.index.name = 'Region'\n",
    "df_10kb = df_10kb.dropna()\n",
    "\n",
    "df_10kb_normalized = (df_10kb - df_10kb.mean()) / df_10kb.std()\n",
    "pca = PCA(n_components=2)\n",
    "pca.fit(df_10kb_normalized.T)\n",
    "transformed_data = pca.transform(df_10kb_normalized.T)\n",
    "\n",
    "labels = ['AK1', 'HG002', 'iPSC', 'NPC', 'H1', 'HUES64']\n",
    "fig, ax = plt.subplots(figsize=(7, 7), constrained_layout=True)\n",
    "for i in range(len(transformed_data)):\n",
    "    ax.scatter(transformed_data[i][0], transformed_data[i][1],\n",
    "               color = sample_palette[df_10kb_normalized.columns[i]],\n",
    "               s = 200)\n",
    "    ax.annotate(labels[i],\n",
    "                (transformed_data[i][0], transformed_data[i][1]), \n",
    "                xytext = (5,5), \n",
    "                textcoords = 'offset points')\n",
    "\n",
    "ax.set_xlabel(f'PC1 ({pca.explained_variance_ratio_[0]*100:.2f}%)')\n",
    "ax.set_ylabel(f'PC2 ({pca.explained_variance_ratio_[1]*100:.2f}%)')\n",
    "\n",
    "plt.savefig('PCA_mCG_10kb.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Random 10kb mCG outside PMD (Supplementary Figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir='/mnt/data/Projects/phenomata/01.Projects/13.AK1_PacBio/01.DNA/Analysis_2023/10kb_methylation/'\n",
    "\n",
    "ak1_random10k = pd.read_table(f\"{dir}AK1_10kb_random10k_met.tab\", index_col=0, names=['AK1'], header=None)\n",
    "hg002_random10k = pd.read_table(f\"{dir}HG002_10kb_random10k_met.tab\", names=['HG002'], index_col=0)\n",
    "ipsc_random10k = pd.read_table(f\"{dir}iPSC_10kb_random10k_met.tab\", names=['iPSC'], index_col=0)\n",
    "h1_random10k = pd.read_table(f\"{dir}H1_10kb_random10k_met.tab\", names=['H1'], index_col=0)\n",
    "hues64_random10k = pd.read_table(f\"{dir}HUES64_10kb_random10k_met.tab\", names=['HUES64'], index_col=0)\n",
    "npc_random10k = pd.read_table(f\"{dir}NPC_10kb_random10k_met.tab\", names=['NPC'], index_col=0)\n",
    "merge_random10k = pd.concat([ak1_random10k, hg002_random10k, ipsc_random10k, h1_random10k, hues64_random10k, npc_random10k], axis=1)\n",
    "\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "fig, ax = plt.subplots(squeeze=False, figsize=(10, 11))\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(merge_random10k.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None,\n",
    "    fmt = \"\",\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    merge_random10k,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None, \n",
    "    fmt = \"\",\n",
    "    cmap = cmap,\n",
    "    linewidth = 0.001, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100)\n",
    "ax[0, 0].set_ylabel('Randomly selected 10kb tiles')\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig(f'{dir}Randomly_Selected_10kbtiles_outside_PMD_mCG_Heatmap_wHG002.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CpG Density of PMD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''/mnt/mone/Project/AK1_PacBio/01.DNA/Analysis_Samples_Merged/DNA_Methylation_Analysis/PMD/CpG_Density\n",
    "bedtools intersect -v -a hg38_primary_chrX_10kb.bed -b hg38.analysisSet_N.bed > hg38_primary_chrX_10kb_woN.bed\n",
    "bedtools intersect -a hg38_primary_chrX_10kb_woN.bed -b hg38_primary_chrX_CpG.bed -c > hg38_primary_chrX_10kb_woN_CpGcount.bed\n",
    "bedtools intersect -a hg38_primary_chrX_10kb_woN_CpGcount.bed -b AK1_PMD_HiFi.bed > 10kb_CpG_number_wAK1PMD.bed\n",
    "bedtools intersect -v -a hg38_primary_chrX_10kb_woN_CpGcount.bed -b AK1_PMD_HiFi.bed > 10kb_CpG_number_woAK1PMD.bed\n",
    "'''\n",
    "\n",
    "cpg_wPMD = pd.read_table(\"10kb_CpG_number_wAK1PMD.bed\", header=None)\n",
    "cpg_woPMD = pd.read_table(\"10kb_CpG_number_woAK1PMD.bed\", header=None)\n",
    "\n",
    "cpg_wPMD = cpg_wPMD[[3]]\n",
    "cpg_woPMD = cpg_woPMD[[3]]\n",
    "\n",
    "cpg_wPMD['PMD'] = 'PMD'\n",
    "cpg_woPMD['PMD'] = 'Non-PMD'\n",
    "\n",
    "cpg_density = pd.concat([cpg_wPMD, cpg_woPMD], axis=0)\n",
    "cpg_density['Normalized_CpG'] = (cpg_density.iloc[:, 0] - cpg_density.iloc[:, 0].min()) / ( cpg_density.iloc[:, 0].max() - cpg_density.iloc[:, 0].min())\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 5), constrained_layout=True)\n",
    "sns.violinplot(data=cpg_density, x='PMD', y='Normalized_CpG', palette={'PMD': 'darkkhaki', 'Non-PMD': 'black'}, ax = ax, cut=0)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('Normalized CpG Density of 10kb tiles')\n",
    "plt.savefig('Normalized_CpG_Density_PMD_Non-PMD.pdf')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PMD from AK1 aligned onto CHM13 genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To see which regions of PMD are not covered by H1 and HUES64\n",
    "dir = '/mnt/data/Projects/phenomata/01.Projects/13.AK1_PacBio/01.DNA/CHM13/'\n",
    "\n",
    "ak1_chm13_pmd = pd.read_table(f\"{dir}AK1_chm13_AK1_chm13_HiFi_PMD_met.txt\", index_col=0)\n",
    "ipsc_chm13_pmd = pd.read_table(f\"{dir}iPSC_chm13_AK1_chm13_HiFi_PMD_met.txt\", index_col=0)\n",
    "h1_chm13_pmd = pd.read_table(f\"{dir}H1_AK1_chm13_HiFi_PMD_met.txt\", index_col=0)\n",
    "hues64_chm13_pmd = pd.read_table(f\"{dir}HUES64_AK1_chm13_HiFi_PMD_met.txt\", index_col=0)\n",
    "npc_chm13_pmd = pd.read_table(f\"{dir}NPC_chm13_AK1_chm13_HiFi_PMD_met.txt\", index_col=0)\n",
    "merge_chm13_pmd = pd.concat([ak1, ipsc, h1, hues64, npc], axis=1)\n",
    "\n",
    "merge = merge[~merge.index.str.startswith('chrY')]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "2022.10.undefined"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
