{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap\n",
    "from sklearn.decomposition import PCA\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "\n",
    "#cmap = matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#104e8b\", \"#ffdab9\", \"#8b0a50\"])\n",
    "cmap = matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#FFFF00\", \"#000000\", \"#0066CC\"])\n",
    "\n",
    "sns.set_theme(font=\"Arial\", font_scale=1.15, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n",
    "%matplotlib\n",
    "%autoindent\n",
    "\n",
    "hap_sample_palette = {'AK1_Hap1': '#CC0033', \n",
    "                      'AK1_Hap2': '#FF0033',\n",
    "                      'AK1_Aggregate': '#FF0000', \n",
    "                      'AK1_WGBS': '#FF00FF',\n",
    "                      'HG002_Hap1': '#999999',\n",
    "                      'HG002_Hap2': '#CCCCCC',\n",
    "                      'HG002_Aggregate': '#666666',\n",
    "                      'iPSC_Hap1': '#0066CC', \n",
    "                      'iPSC_Hap2': '#0000CC', \n",
    "                      'iPSC_Aggregate': '#154360',\n",
    "                      'NPC_Hap1': '#33CC00', \n",
    "                      'NPC_Hap2': '#66FF66', \n",
    "                      'NPC_Aggregate': '#229954',\n",
    "                      'H1': '#40E0D0',\n",
    "                      'HUES64': '#437299'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Haplotype DMR Methylation Difference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_AK1_Hap = pd.read_table(\"DMRs_DSS_AK1_Hap_ALL.bed\", header=None)\n",
    "#df_AK1_Hap = df_AK1_Hap.copy()\n",
    "df_AK1_Hap['mCG_HapDiff'] = df_AK1_Hap.iloc[:, 3].str.split('/').str[1].str.rstrip(')').astype(float) * 100\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6,4), constrained_layout=True)\n",
    "hist = sns.histplot(data=df_AK1_Hap, x='mCG_HapDiff', kde=True, bins=100, color='#664E00', stat='count', ax=ax)\n",
    "ax.set_xlabel(\"Difference in DNA methylation between AK1 Haplotypes (%)\")\n",
    "ax.set_ylabel(\"The Number of AK1 Haplotype DMRs\")\n",
    "ax.axvline(65, color='r', linestyle='--')\n",
    "ax.axvline(-65, color='b', linestyle='--')\n",
    "ax.text(70, ax.get_ylim()[1] * 0.9, '65%', color='r')\n",
    "ax.text(-85, ax.get_ylim()[1] * 0.9, '-65%', color='b')\n",
    "\n",
    "plt.savefig('Histogram_AK1_Haplotype_Methylation_Difference.pdf')\n",
    "plt.clf()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variants Inside Candidate AK1 ICR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_candidateICR = pd.read_table(\"Variants_inside_CandidateICR.tab\", index_col=0)\n",
    "var_candidateICR['ICR_region'] = var_candidateICR.index.str.split(';').str[0]\n",
    "var_candidateICR['Name'] = var_candidateICR.index.str.split(';').str[1]\n",
    "var_candidateICR['PhasedHet'].describe()\n",
    "\n",
    "\n",
    "PhasedHetCategories = [\n",
    "    var_candidateICR['PhasedHet'] == 0,\n",
    "    var_candidateICR['PhasedHet'] == 1,\n",
    "    var_candidateICR['PhasedHet'] == 2,\n",
    "    var_candidateICR['PhasedHet'] == 3,\n",
    "    (var_candidateICR['PhasedHet'] > 3) & (var_candidateICR['PhasedHet'] < 20),\n",
    "    var_candidateICR['PhasedHet'] >= 20\n",
    "    ]\n",
    "PhasedHetCategories_Names = ['0', '1', '2', '3', '4-19', '20+']\n",
    "\n",
    "var_candidateICR['PhasedHetCategories'] = np.select(PhasedHetCategories, PhasedHetCategories_Names, default='Other')\n",
    "\n",
    "PhasedHetCategories_percentage = var_candidateICR['PhasedHetCategories'].value_counts(normalize=True) * 100\n",
    "PhasedHetCategories_percentage_ordered = PhasedHetCategories_percentage.reindex(PhasedHetCategories_Names)\n",
    "cat_starts = PhasedHetCategories_percentage_ordered.cumsum() - PhasedHetCategories_percentage\n",
    "\n",
    "sns.set_theme(font=\"Arial\", font_scale=1.3, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n",
    "\n",
    "colors = LinearSegmentedColormap.from_list(\n",
    "    'white_to_dark_red', ['#ffffff', '#ff0000']\n",
    ")\n",
    "color_range = colors(np.linspace(0,1, len(PhasedHetCategories_percentage_ordered)))[::-1]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(9,1.5), constrained_layout=True)\n",
    "for i, (cat, prop) in enumerate(PhasedHetCategories_percentage_ordered.iteritems()):\n",
    "    ax.barh(y=0, width=prop, left=cat_starts[cat], color=color_range[i], edgecolor='black')\n",
    "\n",
    "for i, (cat, prop) in enumerate(PhasedHetCategories_percentage_ordered.iteritems()):\n",
    "    ax.text(x=cat_starts[cat] + prop / 2, y=0, s=str(cat), ha='center', va='center')\n",
    "    \n",
    "ax.set_yticks([])\n",
    "ax.set_xlim(0,100.05)\n",
    "ax.set_xlabel('Percentage of phased heterozygous variants in Candidate AK1-ICR')\n",
    "\n",
    "plt.savefig(\"Percentage_of_phased_hetvars_in_candidateAK1ICR.pdf\")\n",
    "plt.clf()\n",
    "\n",
    "sns.set_theme(font=\"Arial\", font_scale=1.15, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mCG inside Known ICRs from Imprintome (w/HG002)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_known = pd.read_table(\"Known_ICR_cleaned.bed\", header=None)\n",
    "df_known['ICR_region'] = df_known.iloc[:, 0] + ':' + df_known.iloc[:, 1].astype(str) + '-' + df_known.iloc[:, 2].astype(str)\n",
    "df_known = df_known.set_index('ICR_region')\n",
    "df_known = df_known[[3]]\n",
    "df_known.columns = ['Name']\n",
    "\n",
    "df_ak1_hap1 = pd.read_table(\"AK1_hap1_KnownICR.tab\", index_col=0, header=None)\n",
    "df_ak1_hap2 = pd.read_table(\"AK1_hap2_KnownICR.tab\", index_col=0, header=None)\n",
    "df_ak1 = pd.concat([df_ak1_hap1, df_ak1_hap2], axis=1)\n",
    "df_ak1.index.name = 'ICR_region'\n",
    "df_ak1.columns = ['AK1_Hap1', 'AK1_Hap2']\n",
    "\n",
    "df_hg002_hap1 = pd.read_table(\"HG002_hap1_KnownICR.tab\", index_col=0, header=None)\n",
    "df_hg002_hap2 = pd.read_table(\"HG002_hap2_KnownICR.tab\", index_col=0, header=None)\n",
    "df_hg002 = pd.concat([df_hg002_hap1, df_hg002_hap2], axis=1)\n",
    "df_hg002.index.name = 'ICR_region'\n",
    "df_hg002.columns = ['HG002_Hap1', 'HG002_Hap2']\n",
    "\n",
    "df_ipsc_hap1 = pd.read_table(\"iPSC_hap1_KnownICR.tab\", index_col=0, header=None)\n",
    "df_ipsc_hap2 = pd.read_table(\"iPSC_hap2_KnownICR.tab\", index_col=0, header=None)\n",
    "df_ipsc = pd.concat([df_ipsc_hap1, df_ipsc_hap2], axis=1)\n",
    "df_ipsc.index.name = 'ICR_region'\n",
    "df_ipsc.columns = ['iPSC_Hap1', 'iPSC_Hap2']\n",
    "\n",
    "df_npc_hap1 = pd.read_table(\"NPC_hap1_KnownICR.tab\", index_col=0, header=None)\n",
    "df_npc_hap2 = pd.read_table(\"NPC_hap2_KnownICR.tab\", index_col=0, header=None)\n",
    "df_npc = pd.concat([df_npc_hap1, df_npc_hap2], axis=1)\n",
    "df_npc.index.name = 'ICR_region'\n",
    "df_npc.columns = ['NPC_Hap1', 'NPC_Hap2']\n",
    "\n",
    "df_h1 = pd.read_table(\"H1_MAPQ10_HapMerged_KnownICR.tab\", index_col=0, header=None)\n",
    "df_h1.index.name = 'ICR_region'\n",
    "df_h1.columns = ['H1']\n",
    "\n",
    "df_hues64 = pd.read_table(\"HUES64_MAPQ10_HapMerged_KnownICR.tab\", index_col=0, header=None)\n",
    "df_hues64.index.name = 'ICR_region'\n",
    "df_hues64.columns = ['HUES64']\n",
    "\n",
    "df_known_met = pd.concat([df_known, df_ak1, df_hg002, df_ipsc, df_npc, df_h1, df_hues64], axis=1)\n",
    "df_known_met = df_known_met.iloc[np.argsort(df_known_met['Name'].apply(lambda x: int(x.split('_')[-1])))]\n",
    "df_known_met_hm = df_known_met.iloc[:, 1:]\n",
    "df_known_met_hm['H1_HUES64_mean'] = df_known_met_hm[['H1', 'HUES64']].mean(axis=1)\n",
    "df_known_met_hm = df_known_met_hm.sort_values(by='H1_HUES64_mean', ascending=False)\n",
    "del df_known_met_hm['H1_HUES64_mean']\n",
    "\n",
    "annot = df_known_met_hm.applymap(lambda f: f'{f: .1f}')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8.5, 5), squeeze=False, constrained_layout=True)\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(df_known_met_hm.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    yticklabels = True,\n",
    "    annot = np.full_like(df_known_met_hm, \"NA\", dtype = object),\n",
    "    fmt = \"\",\n",
    "    annot_kws = {\"size\": 10, \"va\": \"center_baseline\", \"color\": \"black\"},\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    df_known_met_hm,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = True,\n",
    "    annot = annot, \n",
    "    fmt = \"\",\n",
    "    annot_kws = {\"size\": 10, \"va\": \"center_baseline\"},\n",
    "    cmap = cmap,\n",
    "    linewidth = 0.5, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100)\n",
    "ax[0, 0].xaxis.tick_top()\n",
    "ax[0, 0].set_xticklabels(['AK1\\nHap1', 'AK1\\nHap2', 'HG002\\nHap1', 'HG002\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64'], rotation=0)\n",
    "ax[0, 0].set_yticklabels(ax[0, 0].get_yticklabels(), fontsize=7)\n",
    "ax[0, 0].yaxis.set_tick_params(width=0)\n",
    "ax[0, 0].set_ylabel('Known ICR from Imprintome (N = 25)')\n",
    "ax[0, 0].yaxis.set_tick_params(width=0, length=0)\n",
    "ax[0, 0].xaxis.set_tick_params(width=0, length=0)\n",
    "heatmap1.set_xlim(0, df_known_met_hm.shape[1] + 0.1)\n",
    "heatmap1.set_ylim(df_known_met_hm.shape[0] + 0.1, 0)\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig('Heatmap_mCG_KnownICRs.png')\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mCG inside Known ICRs from Imprintome (wo/HG002)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_known = pd.read_table(\"Known_ICR_cleaned.bed\", header=None)\n",
    "df_known['ICR_region'] = df_known.iloc[:, 0] + ':' + df_known.iloc[:, 1].astype(str) + '-' + df_known.iloc[:, 2].astype(str)\n",
    "df_known = df_known.set_index('ICR_region')\n",
    "df_known = df_known[[3]]\n",
    "df_known.columns = ['Name']\n",
    "\n",
    "df_known_met_hm = df_known_met.iloc[:, [1, 2, 5, 6, 7, 8, 9, 10]]\n",
    "df_known_met_hm['H1_HUES64_mean'] = df_known_met_hm[['H1', 'HUES64']].mean(axis=1)\n",
    "df_known_met_hm = df_known_met_hm.sort_values(by='H1_HUES64_mean', ascending=False)\n",
    "del df_known_met_hm['H1_HUES64_mean']\n",
    "\n",
    "annot = df_known_met_hm.applymap(lambda f: f'{f: .1f}')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6.5, 5), squeeze=False, constrained_layout=True)\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(df_known_met_hm.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    yticklabels = True,\n",
    "    annot = np.full_like(df_known_met_hm, \"NA\", dtype = object),\n",
    "    fmt = \"\",\n",
    "    annot_kws = {\"size\": 10, \"va\": \"center_baseline\", \"color\": \"black\"},\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    df_known_met_hm,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = True,\n",
    "    annot = annot, \n",
    "    fmt = \"\",\n",
    "    annot_kws = {\"size\": 10, \"va\": \"center_baseline\"},\n",
    "    cmap = cmap,\n",
    "    linewidth = 0.5, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100)\n",
    "ax[0, 0].xaxis.tick_top()\n",
    "ax[0, 0].set_xticklabels(['AK1\\nHap1', 'AK1\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64'], rotation=0)\n",
    "ax[0, 0].set_yticklabels(ax[0, 0].get_yticklabels(), fontsize=7)\n",
    "ax[0, 0].yaxis.set_tick_params(width=0)\n",
    "ax[0, 0].set_ylabel('Known ICR from Imprintome (N = 25)')\n",
    "ax[0, 0].yaxis.set_tick_params(width=0, length=0)\n",
    "ax[0, 0].xaxis.set_tick_params(width=0, length=0)\n",
    "heatmap1.set_xlim(0, df_known_met_hm.shape[1] + 0.1)\n",
    "heatmap1.set_ylim(df_known_met_hm.shape[0] + 0.1, 0)\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig('Heatmap_mCG_KnownICRs_woHG002.png')\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mCG inside Putative ICRs from Imprintome (w/ HG002)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_putative = pd.read_table(\"Putative_ICR_wochrX.bed\", header=None)\n",
    "df_putative['ICR_region'] = df_putative.iloc[:, 0] + ':' + df_putative.iloc[:, 1].astype(str) + '-' + df_putative.iloc[:, 2].astype(str)\n",
    "df_putative = df_putative.set_index('ICR_region')\n",
    "df_putative = df_putative[[3]]\n",
    "df_putative.columns = ['Name']\n",
    "\n",
    "df_ak1_hap1 = pd.read_table(\"AK1_hap1_PutativeICR.tab\", index_col=0, header=None)\n",
    "df_ak1_hap2 = pd.read_table(\"AK1_hap2_PutativeICR.tab\", index_col=0, header=None)\n",
    "df_ak1 = pd.concat([df_ak1_hap1, df_ak1_hap2], axis=1)\n",
    "df_ak1.index.name = 'ICR_region'\n",
    "df_ak1.columns = ['AK1_Hap1', 'AK1_Hap2']\n",
    "\n",
    "df_hg002_hap1 = pd.read_table(\"HG002_hap1_PutativeICR.tab\", index_col=0, header=None)\n",
    "df_hg002_hap2 = pd.read_table(\"HG002_hap2_PutativeICR.tab\", index_col=0, header=None)\n",
    "df_hg002 = pd.concat([df_hg002_hap1, df_hg002_hap2], axis=1)\n",
    "df_hg002.index.name = 'ICR_region'\n",
    "df_hg002.columns = ['HG002_Hap1', 'HG002_Hap2']\n",
    "\n",
    "df_ipsc_hap1 = pd.read_table(\"iPSC_hap1_PutativeICR.tab\", index_col=0, header=None)\n",
    "df_ipsc_hap2 = pd.read_table(\"iPSC_hap2_PutativeICR.tab\", index_col=0, header=None)\n",
    "df_ipsc = pd.concat([df_ipsc_hap1, df_ipsc_hap2], axis=1)\n",
    "df_ipsc.index.name = 'ICR_region'\n",
    "df_ipsc.columns = ['iPSC_Hap1', 'iPSC_Hap2']\n",
    "\n",
    "df_npc_hap1 = pd.read_table(\"NPC_hap1_PutativeICR.tab\", index_col=0, header=None)\n",
    "df_npc_hap2 = pd.read_table(\"NPC_hap2_PutativeICR.tab\", index_col=0, header=None)\n",
    "df_npc = pd.concat([df_npc_hap1, df_npc_hap2], axis=1)\n",
    "df_npc.index.name = 'ICR_region'\n",
    "df_npc.columns = ['NPC_Hap1', 'NPC_Hap2']\n",
    "\n",
    "df_h1 = pd.read_table(\"H1_MAPQ10_HapMerged_PutativeICR.tab\", index_col=0, header=None)\n",
    "df_h1.index.name = 'ICR_region'\n",
    "df_h1.columns = ['H1']\n",
    "\n",
    "df_hues64 = pd.read_table(\"HUES64_MAPQ10_HapMerged_PutativeICR.tab\", index_col=0, header=None)\n",
    "df_hues64.index.name = 'ICR_region'\n",
    "df_hues64.columns = ['HUES64']\n",
    "\n",
    "df_putative_met = pd.concat([df_putative, df_ak1, df_hg002, df_ipsc, df_npc, df_h1, df_hues64], axis=1)\n",
    "df_putative_met = df_putative_met.iloc[np.argsort(df_putative_met['Name'].apply(lambda x: int(x.split('_')[-1])))]\n",
    "df_putative_met_hm = df_putative_met.iloc[:, 1:]\n",
    "df_putative_met_hm['H1_HUES64_mean'] = df_putative_met_hm[['H1', 'HUES64']].mean(axis=1)\n",
    "df_putative_met_hm = df_putative_met_hm.sort_values(by='H1_HUES64_mean', ascending=False)\n",
    "del df_putative_met_hm['H1_HUES64_mean']\n",
    "\n",
    "annot = df_putative_met_hm.applymap(lambda f: f'{f: .1f}')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(7.5, 5), squeeze=False, constrained_layout=True)\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(df_putative_met_hm.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None,\n",
    "    fmt = \"\",\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    df_putative_met_hm,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None, \n",
    "    fmt = \"\",\n",
    "    cmap = cmap,\n",
    "    linewidth = 0.001, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100)\n",
    "ax[0, 0].xaxis.tick_top()\n",
    "ax[0, 0].set_xticklabels(['AK1\\nHap1', 'AK1\\nHap2', 'HG002\\nHap1', 'HG002\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64'], rotation=0)\n",
    "ax[0, 0].set_ylabel('Putative ICR from Imprintome (N = 1,390)')\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig('Heatmap_mCG_PutativeICRs.png')\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mCG inside Putative ICRs from Imprintome (wo/ HG002)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "df_putative_met_hm = df_putative_met_hm.iloc[:, [0, 1, 4, 5, 6, 7, 8, 9]]\n",
    "\n",
    "fig, ax = plt.subplots(squeeze=False, constrained_layout=True)\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(df_putative_met_hm.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None,\n",
    "    fmt = \"\",\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    df_putative_met_hm,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None, \n",
    "    fmt = \"\",\n",
    "    cmap = \"coolwarm\",\n",
    "    linewidth = 0.001, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100)\n",
    "ax[0, 0].xaxis.tick_top()\n",
    "ax[0, 0].set_ylabel('Putative ICR from Imprintome (N = 1,390)')\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig('Heatmap_mCG_PutativeICRs_woHG002.png')\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic statistics inside Candidate ICRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minMax_normalization(series):\n",
    "    return ( series - series.min() ) / ( series.max() - series.min() )\n",
    "\n",
    "\n",
    "df_AK1_candidateICR = pd.read_table(\"Candidate_ICR_metcutoff65_from_AK1_Hap_sorted.bed\", header=None)\n",
    "df_AK1_candidateICR['CpG_Count'] = df_AK1_candidateICR.iloc[:, 3].str.split('(').str[1].str.split('/').str[0].astype(int)\n",
    "df_AK1_candidateICR['ICR_region'] = df_AK1_candidateICR.iloc[:, 0] + ':' + df_AK1_candidateICR.iloc[:, 1].astype(str) + '-' + df_AK1_candidateICR.iloc[:, 2].astype(str)\n",
    "df_AK1_candidateICR['Length'] = (df_AK1_candidateICR.iloc[:, 2] - df_AK1_candidateICR.iloc[:, 1])\n",
    "df_AK1_candidateICR['Norm_CpG_Density'] = minMax_normalization((df_AK1_candidateICR['CpG_Count'] / df_AK1_candidateICR['Length']))\n",
    "df_AK1_candidateICR['mCG_HapDiff'] = df_AK1_candidateICR.iloc[:, 3].str.split('/').str[1].str.rstrip(')').astype(float)\n",
    "sns.scatterplot(data=df_AK1_candidateICR, x='Length', y='Norm_CpG_Density', alpha=0.3)\n",
    "\n",
    "\n",
    "df_AK1_candidateICR = df_AK1_candidateICR.set_index('ICR_region')\n",
    "df_AK1_candidateICR = df_AK1_candidateICR[[3]]\n",
    "df_AK1_candidateICR.columns = ['Name']\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variants inside Candidate ICR in AK1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mCG inside AK1 candidate ICRs (wo/ HG002)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_AK1_candidateICR = pd.read_table(\"Candidate_ICR_metcutoff65_from_AK1_Hap_sorted.bed\", header=None)\n",
    "df_AK1_candidateICR['ICR_region'] = df_AK1_candidateICR.iloc[:, 0] + ':' + df_AK1_candidateICR.iloc[:, 1].astype(str) + '-' + df_AK1_candidateICR.iloc[:, 2].astype(str)\n",
    "df_AK1_candidateICR = df_AK1_candidateICR.set_index('ICR_region')\n",
    "df_AK1_candidateICR = df_AK1_candidateICR[[3]]\n",
    "df_AK1_candidateICR.columns = ['Name']\n",
    "\n",
    "df_ak1ICR_hap1 = pd.read_table(\"AK1_hap1_AK1_candidate_ICR.tab\", index_col=0, header=None)\n",
    "df_ak1ICR_hap2 = pd.read_table(\"AK1_hap2_AK1_candidate_ICR.tab\", index_col=0, header=None)\n",
    "df_ak1ICR = pd.concat([df_ak1ICR_hap1, df_ak1ICR_hap2], axis=1)\n",
    "df_ak1ICR.index.name = 'ICR_region'\n",
    "df_ak1ICR.columns = ['AK1_Hap1', 'AK1_Hap2']\n",
    "\n",
    "df_hg002ICR_hap1 = pd.read_table(\"HG002_hap1_AK1_candidate_ICR.tab\", index_col=0, header=None)\n",
    "df_hg002ICR_hap2 = pd.read_table(\"HG002_hap2_AK1_candidate_ICR.tab\", index_col=0, header=None)\n",
    "df_hg002ICR = pd.concat([df_hg002ICR_hap1, df_hg002ICR_hap2], axis=1)\n",
    "df_hg002ICR.index.name = 'ICR_region'\n",
    "df_hg002ICR.columns = ['HG002_Hap1', 'HG002_Hap2']\n",
    "\n",
    "df_ipscICR_hap1 = pd.read_table(\"iPSC_hap1_AK1_candidate_ICR.tab\", index_col=0, header=None)\n",
    "df_ipscICR_hap2 = pd.read_table(\"iPSC_hap2_AK1_candidate_ICR.tab\", index_col=0, header=None)\n",
    "df_ipscICR = pd.concat([df_ipscICR_hap1, df_ipscICR_hap2], axis=1)\n",
    "df_ipscICR.index.name = 'ICR_region'\n",
    "df_ipscICR.columns = ['iPSC_Hap1', 'iPSC_Hap2']\n",
    "\n",
    "df_npcICR_hap1 = pd.read_table(\"NPC_hap1_AK1_candidate_ICR.tab\", index_col=0, header=None)\n",
    "df_npcICR_hap2 = pd.read_table(\"NPC_hap2_AK1_candidate_ICR.tab\", index_col=0, header=None)\n",
    "df_npcICR = pd.concat([df_npcICR_hap1, df_npcICR_hap2], axis=1)\n",
    "df_npcICR.index.name = 'ICR_region'\n",
    "df_npcICR.columns = ['NPC_Hap1', 'NPC_Hap2']\n",
    "\n",
    "df_h1ICR = pd.read_table(\"H1_MAPQ10_HapMerged_AK1_candidate_ICR.tab\", index_col=0, header=None)\n",
    "df_h1ICR.index.name = 'ICR_region'\n",
    "df_h1ICR.columns = ['H1']\n",
    "\n",
    "df_hues64ICR = pd.read_table(\"HUES64_MAPQ10_HapMerged_AK1_candidate_ICR.tab\", index_col=0, header=None)\n",
    "df_hues64ICR.index.name = 'ICR_region'\n",
    "df_hues64ICR.columns = ['HUES64']\n",
    "\n",
    "# AK1 ICR length distribution\n",
    "df_AK1_candidateICR_length = df_AK1_candidateICR.copy()\n",
    "df_AK1_candidateICR_length['ICR_length'] = df_AK1_candidateICR_length.index.to_series().str.split(':').str[-1].str.split('-').apply(lambda x: int(x[1]) - int(x[0]))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6,4), constrained_layout=True)\n",
    "hist = sns.histplot(data=df_AK1_candidateICR_length, x='ICR_length', kde=True, stat='count', ax=ax)\n",
    "ax.set_xlabel(\"Length of AK1 Candidate ICR\")\n",
    "ax.set_ylabel(\"The Number of AK1 Candidate ICR\")\n",
    "ax.get_xaxis().set_major_formatter(matplotlib.ticker.FuncFormatter(lambda x, p: format(int(x), ',')))\n",
    "ax.axvline(df_AK1_candidateICR_length['ICR_length'].mean(), color='r', linestyle='--')\n",
    "ax.text(df_AK1_candidateICR_length['ICR_length'].mean() + 100, ax.get_ylim()[1] * 0.9, f'Average length: {int(df_AK1_candidateICR_length[\"ICR_length\"].mean())} bp', color='r')\n",
    "\n",
    "plt.savefig('AK1_CandidateICR_Length_Distribution.pdf')\n",
    "\n",
    "# Heatmaps\n",
    "df_AK1_candidateICR_met = pd.concat([df_AK1_candidateICR, df_ak1ICR, df_ipscICR, df_npcICR, df_h1ICR, df_hues64ICR], axis=1)\n",
    "df_AK1_candidateICR_met_hm = df_AK1_candidateICR_met.iloc[:, 1:]\n",
    "df_AK1_candidateICR_met_hm = df_AK1_candidateICR_met_hm.sort_values(by='AK1_Hap1', ascending=False)\n",
    "df_AK1_candidateICR_met_hm = df_AK1_candidateICR_met_hm.dropna(axis=0, subset=['AK1_Hap1'])\n",
    "\n",
    "annot = df_AK1_candidateICR_met_hm.applymap(lambda f: f'{f: .1f}')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(11,8), squeeze=False, constrained_layout=True)\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(df_AK1_candidateICR_met_hm.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None,\n",
    "    fmt = \"\",\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0,\n",
    "    rasterized = True)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    df_AK1_candidateICR_met_hm,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None, \n",
    "    fmt = \"\",\n",
    "    cmap = cmap,\n",
    "    linewidth = 0.001, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100,\n",
    "    rasterized = True)\n",
    "ax[0, 0].xaxis.tick_top()\n",
    "ax[0, 0].set_xticklabels(['AK1\\nHap1', 'AK1\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64'], rotation=0)\n",
    "ax[0, 0].set_ylabel('Candidate ICRs in AK1 (N = 842)')\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig('Heatmap_mCG_AK1candidateICRs_woHG002.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA plot of mCG inside AK1 candidate ICRs (wo/ HG002)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_AK1_candidateICR_met_pca = df_AK1_candidateICR_met_hm.dropna(axis=0, how='any')\n",
    "\n",
    "df_AK1_candidateICR_met_pca_normalized = (df_AK1_candidateICR_met_pca - df_AK1_candidateICR_met_pca.mean()) / df_AK1_candidateICR_met_pca.std()\n",
    "pca = PCA(n_components=2)\n",
    "pca.fit(df_AK1_candidateICR_met_pca_normalized.T)\n",
    "transformed_data = pca.transform(df_AK1_candidateICR_met_pca_normalized.T)\n",
    "\n",
    "sns.set(font=\"Arial\", font_scale=0.9, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n",
    "labels = ['AK1\\nHap1', 'AK1\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64']\n",
    "fig, ax = plt.subplots(figsize=(7, 7), constrained_layout=True)\n",
    "for i in range(len(transformed_data)):\n",
    "    ax.scatter(transformed_data[i][0], transformed_data[i][1],\n",
    "               color = hap_sample_palette[df_AK1_candidateICR_met_pca.columns[i]],\n",
    "               s = 200)\n",
    "    ax.annotate(labels[i],\n",
    "                (transformed_data[i][0], transformed_data[i][1]), \n",
    "                xytext = (5,5), \n",
    "                textcoords = 'offset points')\n",
    "'''\n",
    "    ax.annotate(df_ak1novel_met_pca.columns[i], \n",
    "                (transformed_data[i][0], transformed_data[i][1]), \n",
    "                xytext = (5,5), \n",
    "                textcoords = 'offset points')\n",
    "'''\n",
    "\n",
    "ax.set_xlabel(f'PC1 ({pca.explained_variance_ratio_[0]*100:.2f}%)')\n",
    "ax.set_ylabel(f'PC2 ({pca.explained_variance_ratio_[1]*100:.2f}%)')\n",
    "sns.set(font=\"Arial\", font_scale=1.15, style='ticks') # Back to as it was\n",
    "plt.savefig('PCA_mCG_AK1candidateICRs_woHG002.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA plot of mCG inside AK1 candidate ICRs - HiFi haps, HiFi combined and WGBS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ak1combICR = pd.read_table(\"AK1_combinedmet.tab\", index_col=0, header=None)\n",
    "df_ak1combICR.index.name = 'ICR_region'\n",
    "df_ak1combICR.columns = ['AK1_Aggregate']\n",
    "\n",
    "df_ipsccombICR = pd.read_table(\"iPSC_combinedmet.tab\", index_col=0, header=None)\n",
    "df_ipsccombICR.index.name = 'ICR_region'\n",
    "df_ipsccombICR.columns = ['iPSC_Aggregate']\n",
    "\n",
    "df_npccombICR = pd.read_table(\"NPC_combinedmet.tab\", index_col=0, header=None)\n",
    "df_npccombICR.index.name = 'ICR_region'\n",
    "df_npccombICR.columns = ['NPC_Aggregate']\n",
    "\n",
    "df_ak1wgbsICR = pd.read_table(\"AK1_WGBS_MAPQ10_HapMerged_AK1_candidate_ICR.tab\", index_col=0, header=None)\n",
    "df_ak1wgbsICR.index.name = 'ICR_region'\n",
    "df_ak1wgbsICR.columns = ['AK1_WGBS']\n",
    "\n",
    "df_hg002combICR = pd.read_table(\"HG002_combinedmet.tab\", index_col=0, header=None)\n",
    "df_hg002combICR.index.name = 'ICR_region'\n",
    "df_hg002combICR.columns = ['HG002_Aggregate']\n",
    "\n",
    "\n",
    "'''\n",
    "df_AK1_w_combined_candidateICR_met = pd.concat([df_AK1_candidateICR, df_ak1ICR, df_ipscICR, df_npcICR, df_h1ICR, df_hues64ICR, df_ak1combICR, df_ipsccombICR, df_npccombICR, df_ak1wgbsICR, df_hg002ICR, df_hg002combICR], axis=1)\n",
    "df_AK1_w_combined_candidateICR_met_hm = df_AK1_w_combined_candidateICR_met.iloc[:, 1:]\n",
    "\n",
    "df_AK1_w_combined_candidateICR_met_pca = df_AK1_w_combined_candidateICR_met_hm.dropna(axis=0, how='any')\n",
    "df_AK1_w_combined_candidateICR_met_pca_normalized = (df_AK1_w_combined_candidateICR_met_pca - df_AK1_w_combined_candidateICR_met_pca.mean()) / df_AK1_w_combined_candidateICR_met_pca.std()\n",
    "pca = PCA(n_components=3)\n",
    "pca.fit(df_AK1_w_combined_candidateICR_met_pca_normalized.T)\n",
    "transformed_data = pca.transform(df_AK1_w_combined_candidateICR_met_pca_normalized.T)\n",
    "\n",
    "sns.set_theme(font=\"Arial\", font_scale=1.2, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n",
    "labels = ['AK1\\nHap1', 'AK1\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64', 'AK1\\nAggregate', 'iPSC\\nAggregate', 'NPC\\nAggregate', 'AK1\\nWGBS', 'HG002\\nHap1', 'HG002\\nHap2', 'HG002\\nAggregate']\n",
    "fig, ax = plt.subplots(figsize=(7, 7), constrained_layout=True)\n",
    "for i in range(len(transformed_data)):\n",
    "    ax.scatter(transformed_data[i][0], transformed_data[i][1],\n",
    "               color = hap_sample_palette[df_AK1_w_combined_candidateICR_met_pca.columns[i]],\n",
    "               s = 200)\n",
    "    ax.annotate(labels[i],\n",
    "                (transformed_data[i][0], transformed_data[i][1]), \n",
    "                xytext = (5,5), \n",
    "                textcoords = 'offset points')\n",
    "\"\"\"\n",
    "    ax.annotate(df_ak1novel_met_pca.columns[i], \n",
    "                (transformed_data[i][0], transformed_data[i][1]), \n",
    "                xytext = (5,5), \n",
    "                textcoords = 'offset points')\n",
    "\"\"\"\n",
    "ax.set_xlabel(f'PC1 ({pca.explained_variance_ratio_[0]*100:.2f}%)')\n",
    "ax.set_ylabel(f'PC2 ({pca.explained_variance_ratio_[1]*100:.2f}%)')\n",
    "sns.set_theme(font=\"Arial\", font_scale=1.15, style='ticks') # Back to as it was\n",
    "plt.savefig('PCA_mCG_AK1candidateICRs_woHG002_wAK1WGBS.pdf')\n",
    "\n",
    "'''\n",
    "df_AK1_w_combined_candidateICR_met = pd.concat([df_AK1_candidateICR, df_ak1ICR, df_ipscICR, df_npcICR, df_h1ICR, df_hues64ICR, df_ak1wgbsICR], axis=1)\n",
    "df_AK1_w_combined_candidateICR_met_hm = df_AK1_w_combined_candidateICR_met.iloc[:, 1:]\n",
    "\n",
    "df_AK1_w_combined_candidateICR_met_pca = df_AK1_w_combined_candidateICR_met_hm.dropna(axis=0, how='any')\n",
    "df_AK1_w_combined_candidateICR_met_pca_normalized = (df_AK1_w_combined_candidateICR_met_pca - df_AK1_w_combined_candidateICR_met_pca.mean()) / df_AK1_w_combined_candidateICR_met_pca.std()\n",
    "pca = PCA(n_components=2)\n",
    "pca.fit(df_AK1_w_combined_candidateICR_met_pca_normalized.T)\n",
    "transformed_data = pca.transform(df_AK1_w_combined_candidateICR_met_pca_normalized.T)\n",
    "\n",
    "sns.set_theme(font=\"Arial\", font_scale=1.2, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n",
    "#labels = ['AK1\\nHap1', 'AK1\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64', 'AK1\\nAggregate', 'iPSC\\nAggregate', 'NPC\\nAggregate', 'AK1\\nWGBS']\n",
    "labels = ['AK1\\nHap1', 'AK1\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64', 'AK1\\nWGBS']\n",
    "fig, ax = plt.subplots(figsize=(7, 7), constrained_layout=True)\n",
    "for i in range(len(transformed_data)):\n",
    "    ax.scatter(transformed_data[i][0], transformed_data[i][1],\n",
    "               color = hap_sample_palette[df_AK1_w_combined_candidateICR_met_pca.columns[i]],\n",
    "               s = 200)\n",
    "    ax.annotate(labels[i],\n",
    "                (transformed_data[i][0], transformed_data[i][1]), \n",
    "                xytext = (5,5), \n",
    "                textcoords = 'offset points')\n",
    "'''\n",
    "    ax.annotate(df_ak1novel_met_pca.columns[i], \n",
    "                (transformed_data[i][0], transformed_data[i][1]), \n",
    "                xytext = (5,5), \n",
    "                textcoords = 'offset points')\n",
    "'''\n",
    "\n",
    "ax.set_xlabel(f'PC1 ({pca.explained_variance_ratio_[0]*100:.2f}%)')\n",
    "ax.set_ylabel(f'PC2 ({pca.explained_variance_ratio_[1]*100:.2f}%)')\n",
    "sns.set_theme(font=\"Arial\", font_scale=1.15, style='ticks') # Back to as it was\n",
    "plt.savefig('PCA_mCG_AK1candidateICRs_woHG002_wAK1WGBS.pdf')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mCG inside novel AK1 candidate ICRs (w/ HG002)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# /mnt/mone/Project/AK1_PacBio/01.DNA/Analysis_Samples_Merged/DNA_Methylation_Analysis/DMR/Haplotype_DSS/Imprinting/Gene_Association/mCG_Candidate_novelAK1_ICR\n",
    "df_candidate_novelICR = pd.read_table(\"Candidate_ICR_AK1_Gene_Association_NOVEL.bed\", header=None)\n",
    "df_candidate_novelICR['ICR_region'] = df_candidate_novelICR.iloc[:, 0] + ':' + df_candidate_novelICR.iloc[:, 1].astype(str) + '-' + df_candidate_novelICR.iloc[:, 2].astype(str)\n",
    "df_candidate_novelICR = df_candidate_novelICR.set_index('ICR_region')\n",
    "df_candidate_novelICR = df_candidate_novelICR[[3]]\n",
    "df_candidate_novelICR.columns = ['Name']\n",
    "\n",
    "df_ak1_hap1 = pd.read_table(\"AK1_hap1_AK1ICR.tab\", index_col=0, header=None)\n",
    "df_ak1_hap2 = pd.read_table(\"AK1_hap2_AK1ICR.tab\", index_col=0, header=None)\n",
    "df_ak1 = pd.concat([df_ak1_hap1, df_ak1_hap2], axis=1)\n",
    "df_ak1.index.name = 'ICR_region'\n",
    "df_ak1.columns = ['AK1_Hap1', 'AK1_Hap2']\n",
    "\n",
    "df_hg002_hap1 = pd.read_table(\"HG002_hap1_AK1ICR.tab\", index_col=0, header=None)\n",
    "df_hg002_hap2 = pd.read_table(\"HG002_hap2_AK1ICR.tab\", index_col=0, header=None)\n",
    "df_hg002 = pd.concat([df_hg002_hap1, df_hg002_hap2], axis=1)\n",
    "df_hg002.index.name = 'ICR_region'\n",
    "df_hg002.columns = ['HG002_Hap1', 'HG002_Hap2']\n",
    "\n",
    "df_ipsc_hap1 = pd.read_table(\"iPSC_hap1_AK1ICR.tab\", index_col=0, header=None)\n",
    "df_ipsc_hap2 = pd.read_table(\"iPSC_hap2_AK1ICR.tab\", index_col=0, header=None)\n",
    "df_ipsc = pd.concat([df_ipsc_hap1, df_ipsc_hap2], axis=1)\n",
    "df_ipsc.index.name = 'ICR_region'\n",
    "df_ipsc.columns = ['iPSC_Hap1', 'iPSC_Hap2']\n",
    "\n",
    "df_npc_hap1 = pd.read_table(\"NPC_hap1_AK1ICR.tab\", index_col=0, header=None)\n",
    "df_npc_hap2 = pd.read_table(\"NPC_hap2_AK1ICR.tab\", index_col=0, header=None)\n",
    "df_npc = pd.concat([df_npc_hap1, df_npc_hap2], axis=1)\n",
    "df_npc.index.name = 'ICR_region'\n",
    "df_npc.columns = ['NPC_Hap1', 'NPC_Hap2']\n",
    "\n",
    "df_h1 = pd.read_table(\"H1_HapMerged_AK1ICR.tab\", index_col=0, header=None)\n",
    "df_h1.index.name = 'ICR_region'\n",
    "df_h1.columns = ['H1']\n",
    "\n",
    "df_hues64 = pd.read_table(\"HUES64_HapMerged_AK1ICR.tab\", index_col=0, header=None)\n",
    "df_hues64.index.name = 'ICR_region'\n",
    "df_hues64.columns = ['HUES64']\n",
    "\n",
    "'''\n",
    "df_ak1novel_met = pd.concat([df_candidate_novelICR, df_ak1, df_hg002, df_ipsc, df_npc, df_h1, df_hues64], axis=1)\n",
    "df_ak1novel_met_hm = df_ak1novel_met.iloc[:, 1:]\n",
    "df_ak1novel_met_hm['H1_HUES64_mean'] = df_ak1novel_met_hm[['H1', 'HUES64']].mean(axis=1)\n",
    "df_ak1novel_met_hm = df_ak1novel_met_hm.sort_values(by='H1_HUES64_mean', ascending=False)\n",
    "del df_ak1novel_met_hm['H1_HUES64_mean']\n",
    "\n",
    "annot = df_ak1novel_met_hm.applymap(lambda f: f'{f: .1f}')\n",
    "\n",
    "fig, ax = plt.subplots(squeeze=False, constrained_layout=True)\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(df_ak1novel_met_hm.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None,\n",
    "    fmt = \"\",\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    df_ak1novel_met_hm,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None, \n",
    "    fmt = \"\",\n",
    "    cmap = \"coolwarm\",\n",
    "    linewidth = 0.001, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100)\n",
    "ax[0, 0].xaxis.tick_top()\n",
    "ax[0, 0].set_xticklabels(ax[0, 0].get_xticklabels(), rotation=45)\n",
    "ax[0, 0].set_ylabel('Novel ICRs in AK1 (N = 609)')\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig('Heatmap_mCG_AK1novelICRs_1.png')\n",
    "'''\n",
    "\n",
    "df_ak1novel_met = pd.concat([df_candidate_novelICR, df_ak1, df_hg002, df_ipsc, df_npc, df_h1, df_hues64], axis=1)\n",
    "df_ak1novel_met_hm = df_ak1novel_met.iloc[:, 1:]\n",
    "df_ak1novel_met_hm = df_ak1novel_met_hm.sort_values(by='AK1_Hap1', ascending=False)\n",
    "df_ak1novel_met_hm = df_ak1novel_met_hm.dropna(axis=0, subset=['AK1_Hap1'])\n",
    "annot = df_ak1novel_met_hm.applymap(lambda f: f'{f: .1f}')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(7.5, 5), squeeze=False, constrained_layout=True)\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(df_ak1novel_met_hm.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None,\n",
    "    fmt = \"\",\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    df_ak1novel_met_hm,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None, \n",
    "    fmt = \"\",\n",
    "    cmap = cmap,\n",
    "    linewidth = 0.001, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100)\n",
    "ax[0, 0].xaxis.tick_top()\n",
    "ax[0, 0].set_xticklabels(['AK1\\nHap1', 'AK1\\nHap2', 'HG002\\nHap1', 'HG002\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64'], rotation=0)\n",
    "ax[0, 0].set_ylabel('Novel ICRs in AK1 (N = 606)')\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig('Heatmap_mCG_AK1novelICRs.png')\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mCG inside novel AK1 candidate ICRs (wo/ HG002)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ak1novel_met = pd.concat([df_candidate_novelICR, df_ak1, df_hg002, df_ipsc, df_npc, df_h1, df_hues64], axis=1)\n",
    "df_ak1novel_met_hm = df_ak1novel_met.iloc[:, 1:]\n",
    "df_ak1novel_met_hm = df_ak1novel_met_hm.sort_values(by='AK1_Hap1', ascending=False)\n",
    "df_ak1novel_met_hm = df_ak1novel_met_hm.dropna(axis=0, subset=['AK1_Hap1'])\n",
    "df_ak1novel_met_hm = df_ak1novel_met_hm.iloc[:, [0, 1, 4, 5, 6, 7, 8, 9]]\n",
    "\n",
    "annot = df_ak1novel_met_hm.applymap(lambda f: f'{f: .1f}')\n",
    "\n",
    "fig, ax = plt.subplots(squeeze=False, constrained_layout=True)\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(df_ak1novel_met_hm.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None,\n",
    "    fmt = \"\",\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    df_ak1novel_met_hm,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None, \n",
    "    fmt = \"\",\n",
    "    cmap = cmap,\n",
    "    linewidth = 0.001, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100)\n",
    "ax[0, 0].xaxis.tick_top()\n",
    "ax[0, 0].set_xticklabels(ax[0, 0].get_xticklabels(), rotation=45)\n",
    "ax[0, 0].set_ylabel('Novel ICRs in AK1 (N = 606)')\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig('Heatmap_mCG_AK1novelICRs_2_woHG002.png')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove HG002 and dropna\n",
    "df_ak1novel_met_hm = df_ak1novel_met_hm.iloc[:, [0, 1, 4, 5, 6, 7, 8, 9]]\n",
    "df_ak1novel_met_pca = df_ak1novel_met_hm.dropna(axis=0, how='any')\n",
    "\n",
    "df_ak1novel_met_pca_normalized = (df_ak1novel_met_pca - df_ak1novel_met_pca.mean()) / df_ak1novel_met_pca.std()\n",
    "pca = PCA(n_components=2)\n",
    "pca.fit(df_ak1novel_met_pca_normalized.T)\n",
    "transformed_data = pca.transform(df_ak1novel_met_pca_normalized.T)\n",
    "\n",
    "sns.set(font=\"Arial\", font_scale=0.9, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n",
    "labels = ['AK1\\nHap1', 'AK1\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64']\n",
    "fig, ax = plt.subplots(figsize=(7, 7), constrained_layout=True)\n",
    "for i in range(len(transformed_data)):\n",
    "    ax.scatter(transformed_data[i][0], transformed_data[i][1],\n",
    "               color = hap_sample_palette[df_ak1novel_met_pca.columns[i]],\n",
    "               s = 200)\n",
    "    ax.annotate(labels[i],\n",
    "                (transformed_data[i][0], transformed_data[i][1]), \n",
    "                xytext = (5,5), \n",
    "                textcoords = 'offset points')\n",
    "'''\n",
    "    ax.annotate(df_ak1novel_met_pca.columns[i], \n",
    "                (transformed_data[i][0], transformed_data[i][1]), \n",
    "                xytext = (5,5), \n",
    "                textcoords = 'offset points')\n",
    "'''\n",
    "\n",
    "ax.set_xlabel(f'PC1 ({pca.explained_variance_ratio_[0]*100:.2f}%)')\n",
    "ax.set_ylabel(f'PC2 ({pca.explained_variance_ratio_[1]*100:.2f}%)')\n",
    "sns.set(font=\"Arial\", font_scale=1.15, style='ticks') # Back to as it was\n",
    "plt.savefig('PCA_mCG_AK1novelICRs_woHG002.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AK1-ICR retained in iPSC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ICR_retained = pd.read_table(\"iPSC-HapDMR_AK1-ICR_intersect.bed\", header=None)\n",
    "df_ICR_retained['ICR_retained'] = df_ICR_retained.iloc[:, 0] + ':' + df_ICR_retained.iloc[:, 1].astype(str) + '-' + df_ICR_retained.iloc[:, 2].astype(str)\n",
    "df_ICR_retained = df_ICR_retained.set_index('ICR_retained')\n",
    "df_ICR_retained = df_ICR_retained[[3]]\n",
    "df_ICR_retained.columns = ['Name']\n",
    "\n",
    "df_ak1_hap1 = pd.read_table(\"AK1_hap1_iPSC-HapDMR_AK1-ICR.tab\", index_col=0, header=None)\n",
    "df_ak1_hap2 = pd.read_table(\"AK1_hap2_iPSC-HapDMR_AK1-ICR.tab\", index_col=0, header=None)\n",
    "df_ak1 = pd.concat([df_ak1_hap1, df_ak1_hap2], axis=1)\n",
    "df_ak1.index.name = 'ICR_retained'\n",
    "df_ak1.columns = ['AK1_Hap1', 'AK1_Hap2']\n",
    "\n",
    "df_hg002_hap1 = pd.read_table(\"HG002_hap1_iPSC-HapDMR_AK1-ICR.tab\", index_col=0, header=None)\n",
    "df_hg002_hap2 = pd.read_table(\"HG002_hap2_iPSC-HapDMR_AK1-ICR.tab\", index_col=0, header=None)\n",
    "df_hg002 = pd.concat([df_hg002_hap1, df_hg002_hap2], axis=1)\n",
    "df_hg002.index.name = 'ICR_retained'\n",
    "df_hg002.columns = ['HG002_Hap1', 'HG002_Hap2']\n",
    "\n",
    "df_ipsc_hap1 = pd.read_table(\"iPSC_hap1_iPSC-HapDMR_AK1-ICR.tab\", index_col=0, header=None)\n",
    "df_ipsc_hap2 = pd.read_table(\"iPSC_hap2_iPSC-HapDMR_AK1-ICR.tab\", index_col=0, header=None)\n",
    "df_ipsc = pd.concat([df_ipsc_hap1, df_ipsc_hap2], axis=1)\n",
    "df_ipsc.index.name = 'ICR_retained'\n",
    "df_ipsc.columns = ['iPSC_Hap1', 'iPSC_Hap2']\n",
    "\n",
    "df_npc_hap1 = pd.read_table(\"NPC_hap1_iPSC-HapDMR_AK1-ICR.tab\", index_col=0, header=None)\n",
    "df_npc_hap2 = pd.read_table(\"NPC_hap2_iPSC-HapDMR_AK1-ICR.tab\", index_col=0, header=None)\n",
    "df_npc = pd.concat([df_npc_hap1, df_npc_hap2], axis=1)\n",
    "df_npc.index.name = 'ICR_retained'\n",
    "df_npc.columns = ['NPC_Hap1', 'NPC_Hap2']\n",
    "\n",
    "df_h1 = pd.read_table(\"H1_MAPQ10_HapMerged_iPSC-HapDMR_AK1-ICR.tab\", index_col=0, header=None)\n",
    "df_h1.index.name = 'ICR_retained'\n",
    "df_h1.columns = ['H1']\n",
    "\n",
    "df_hues64 = pd.read_table(\"HUES64_MAPQ10_HapMerged_iPSC-HapDMR_AK1-ICR.tab\", index_col=0, header=None)\n",
    "df_hues64.index.name = 'ICR_retained'\n",
    "df_hues64.columns = ['HUES64']\n",
    "\n",
    "\n",
    "df_ICR_retained_met = pd.concat([df_ICR_retained, df_ak1, df_ipsc, df_npc, df_h1, df_hues64], axis=1)\n",
    "df_ICR_retained_met_hm = df_ICR_retained_met.iloc[:, 1:]\n",
    "df_ICR_retained_met_hm = df_ICR_retained_met_hm.sort_values(by='AK1_Hap1', ascending=False)\n",
    "df_ICR_retained_met_hm = df_ICR_retained_met_hm.dropna(axis=0, subset=['AK1_Hap1'])\n",
    "\n",
    "df_ICR_retained_met_hm_1 = df_ICR_retained_met_hm.query('(35 <= H1 <= 65) or (35 <= HUES64 <= 65)')\n",
    "\n",
    "annot = df_ICR_retained_met_hm_1.applymap(lambda f: f'{f: .1f}')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(11, 8), squeeze=False, constrained_layout=True)\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(df_ICR_retained_met_hm_1.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None,\n",
    "    fmt = \"\",\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0, \n",
    "    rasterized = True)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    df_ICR_retained_met_hm_1,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None, \n",
    "    fmt = \"\",\n",
    "    cmap = cmap,\n",
    "    linewidth = 0.001, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100,\n",
    "    rasterized = True)\n",
    "ax[0, 0].xaxis.tick_top()\n",
    "ax[0, 0].set_xticklabels(['AK1\\nHap1', 'AK1\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64'], rotation=0)\n",
    "ax[0, 0].set_ylabel('Candidate AK1 ICR retained in iPSC\\nthat shows intermediate level of mCG in either one of ESCs (N = 23)')\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig('Candidate-AK1-ICR_retained_in_iPSC_but_intermediate_in_ESC_woHG002.pdf')\n",
    "\n",
    "\n",
    "df_ICR_retained_met_hm_2 = df_ICR_retained_met_hm[~df_ICR_retained_met_hm.index.isin(df_ICR_retained_met_hm_1.index)]\n",
    "df_ICR_retained_met_hm_2 = df_ICR_retained_met_hm_2.dropna(subset=['H1', 'HUES64'], how='all')\n",
    "\n",
    "annot = df_ICR_retained_met_hm_2.applymap(lambda f: f'{f: .1f}')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(11, 8), squeeze=False, constrained_layout=True)\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(df_ICR_retained_met_hm_2.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None,\n",
    "    fmt = \"\",\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0, \n",
    "    rasterized = True)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    df_ICR_retained_met_hm_2,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None, \n",
    "    fmt = \"\",\n",
    "    cmap = cmap,\n",
    "    linewidth = 0.001, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100,\n",
    "    rasterized = True)\n",
    "ax[0, 0].xaxis.tick_top()\n",
    "ax[0, 0].set_xticklabels(['AK1\\nHap1', 'AK1\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64'], rotation=0)\n",
    "ax[0, 0].set_ylabel('Candidate AK1 ICR retained in iPSC\\n (Somatic Memory, N = 41)')\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig('Candidate-AK1-ICR_retained_in_iPSC_Somatic_Memory.pdf')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AK1-ICR restored in iPSC (but could be aberrant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ICR_restored = pd.read_table(\"AK1-ICR_notiPSC-DMR_intersect.bed\", header=None)\n",
    "df_ICR_restored['ICR_restored'] = df_ICR_restored.iloc[:, 0] + ':' + df_ICR_restored.iloc[:, 1].astype(str) + '-' + df_ICR_restored.iloc[:, 2].astype(str)\n",
    "df_ICR_restored = df_ICR_restored.set_index('ICR_restored')\n",
    "df_ICR_restored = df_ICR_restored[[3]]\n",
    "df_ICR_restored.columns = ['Name']\n",
    "\n",
    "df_ak1_hap1 = pd.read_table(\"AK1_hap1_AK1-ICR_notiPSC-DMR.tab\", index_col=0, header=None)\n",
    "df_ak1_hap2 = pd.read_table(\"AK1_hap2_AK1-ICR_notiPSC-DMR.tab\", index_col=0, header=None)\n",
    "df_ak1 = pd.concat([df_ak1_hap1, df_ak1_hap2], axis=1)\n",
    "df_ak1.index.name = 'ICR_restored'\n",
    "df_ak1.columns = ['AK1_Hap1', 'AK1_Hap2']\n",
    "\n",
    "df_hg002_hap1 = pd.read_table(\"HG002_hap1_AK1-ICR_notiPSC-DMR.tab\", index_col=0, header=None)\n",
    "df_hg002_hap2 = pd.read_table(\"HG002_hap2_AK1-ICR_notiPSC-DMR.tab\", index_col=0, header=None)\n",
    "df_hg002 = pd.concat([df_hg002_hap1, df_hg002_hap2], axis=1)\n",
    "df_hg002.index.name = 'ICR_restored'\n",
    "df_hg002.columns = ['HG002_Hap1', 'HG002_Hap2']\n",
    "\n",
    "df_ipsc_hap1 = pd.read_table(\"iPSC_hap1_AK1-ICR_notiPSC-DMR.tab\", index_col=0, header=None)\n",
    "df_ipsc_hap2 = pd.read_table(\"iPSC_hap2_AK1-ICR_notiPSC-DMR.tab\", index_col=0, header=None)\n",
    "df_ipsc = pd.concat([df_ipsc_hap1, df_ipsc_hap2], axis=1)\n",
    "df_ipsc.index.name = 'ICR_restored'\n",
    "df_ipsc.columns = ['iPSC_Hap1', 'iPSC_Hap2']\n",
    "\n",
    "df_npc_hap1 = pd.read_table(\"NPC_hap1_AK1-ICR_notiPSC-DMR.tab\", index_col=0, header=None)\n",
    "df_npc_hap2 = pd.read_table(\"NPC_hap2_AK1-ICR_notiPSC-DMR.tab\", index_col=0, header=None)\n",
    "df_npc = pd.concat([df_npc_hap1, df_npc_hap2], axis=1)\n",
    "df_npc.index.name = 'ICR_restored'\n",
    "df_npc.columns = ['NPC_Hap1', 'NPC_Hap2']\n",
    "\n",
    "df_h1 = pd.read_table(\"H1_MAPQ10_HapMerged_AK1-ICR_notiPSC-DMR.tab\", index_col=0, header=None)\n",
    "df_h1.index.name = 'ICR_restored'\n",
    "df_h1.columns = ['H1']\n",
    "\n",
    "df_hues64 = pd.read_table(\"HUES64_MAPQ10_HapMerged_AK1-ICR_notiPSC-DMR.tab\", index_col=0, header=None)\n",
    "df_hues64.index.name = 'ICR_restored'\n",
    "df_hues64.columns = ['HUES64']\n",
    "\n",
    "\n",
    "df_ICR_restored_met = pd.concat([df_ICR_restored, df_ak1, df_ipsc, df_npc, df_h1, df_hues64], axis=1)\n",
    "df_ICR_restored_met_hm = df_ICR_restored_met.iloc[:, 1:]\n",
    "df_ICR_restored_met_hm = df_ICR_restored_met_hm.sort_values(by='AK1_Hap1', ascending=False)\n",
    "df_ICR_restored_met_hm = df_ICR_restored_met_hm.dropna(axis=0, subset=['AK1_Hap1'])\n",
    "\n",
    "df_ICR_restored_met_hm_1 = df_ICR_restored_met_hm.query('(H1 <= 35) or (HUES64 <= 35)')\n",
    "\n",
    "annot = df_ICR_restored_met_hm_1.applymap(lambda f: f'{f: .1f}')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(11, 8), squeeze=False, constrained_layout=True)\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(df_ICR_restored_met_hm_1.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None,\n",
    "    fmt = \"\",\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0, \n",
    "    rasterized = True)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    df_ICR_restored_met_hm_1,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None, \n",
    "    fmt = \"\",\n",
    "    cmap = cmap,\n",
    "    linewidth = 0.001, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100,\n",
    "    rasterized = True)\n",
    "ax[0, 0].xaxis.tick_top()\n",
    "ax[0, 0].set_xticklabels(['AK1\\nHap1', 'AK1\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64'], rotation=0)\n",
    "ax[0, 0].set_ylabel('Candidate AK1 ICR retained in iPSC\\nthat shows intermediate level of mCG in either one of ESCs (N = 23)')\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig('Candidate-AK1-ICR_retained_in_iPSC_but_intermediate_in_ESC_woHG002.pdf')\n",
    "\n",
    "\n",
    "df_ICR_retained_met_hm_2 = df_ICR_retained_met_hm[~df_ICR_retained_met_hm.index.isin(df_ICR_retained_met_hm_1.index)]\n",
    "df_ICR_retained_met_hm_2 = df_ICR_retained_met_hm_2.dropna(subset=['H1', 'HUES64'], how='all')\n",
    "\n",
    "annot = df_ICR_retained_met_hm_2.applymap(lambda f: f'{f: .1f}')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(11, 8), squeeze=False, constrained_layout=True)\n",
    "# For NA values\n",
    "heatmap1 = sns.heatmap(\n",
    "    np.where(df_ICR_retained_met_hm_2.isna(), 0, np.nan),\n",
    "    ax = ax[0, 0],\n",
    "    cbar = False,\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None,\n",
    "    fmt = \"\",\n",
    "    cmap = ListedColormap(['grey']),\n",
    "    linewidth = 0, \n",
    "    rasterized = True)\n",
    "# For non-NA values\n",
    "heatmap2 = sns.heatmap(\n",
    "    df_ICR_retained_met_hm_2,\n",
    "    ax = ax[0, 0],\n",
    "    xticklabels = True,\n",
    "    yticklabels = False,\n",
    "    annot = None, \n",
    "    fmt = \"\",\n",
    "    cmap = cmap,\n",
    "    linewidth = 0.001, \n",
    "    linecolor = \"black\",\n",
    "    vmin = 0,\n",
    "    vmax = 100,\n",
    "    rasterized = True)\n",
    "ax[0, 0].xaxis.tick_top()\n",
    "ax[0, 0].set_xticklabels(['AK1\\nHap1', 'AK1\\nHap2', 'iPSC\\nHap1', 'iPSC\\nHap2', 'NPC\\nHap1', 'NPC\\nHap2', 'H1', 'HUES64'], rotation=0)\n",
    "ax[0, 0].set_ylabel('Candidate AK1 ICR retained in iPSC\\n (Somatic Memory, N = 41)')\n",
    "heatmap2.collections[1].colorbar.set_label(label='DNA methylation (%)', rotation=270, labelpad=10)\n",
    "plt.savefig('Candidate-AK1-ICR_retained_in_iPSC_Somatic_Memory.pdf')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
