{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib ##\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.mixture import GaussianMixture\n",
    "sns.set(font=\"Arial\", font_scale=1.2, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n",
    "import gseapy as gp\n",
    "\n",
    "#%matplotlib inline\n",
    "sample_palette = {'AK1':'#FF0000', 'iPSC':'#154360', 'H1': '#40E0D0', 'HUES64': '#437299', 'NPC':'#229954', 'HG002': '#545454'}\n",
    "cmap = matplotlib.colors.LinearSegmentedColormap.from_list(\"\", [\"#104e8b\", \"#ffdab9\", \"#8b0a50\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gain-Loss counts for AK1-iPSC-NPC relationship"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# /mnt/mone/Project/AK1_PacBio/01.DNA/Analysis_Samples_Merged/DNA_Methylation_Analysis/DMR/DSS/DSS_Outputs/Gain_Loss_Counts\n",
    "# AK1-iPSC DMRs not overlapped with PMD\n",
    "\n",
    "rfh = open(\"mCG_Gain_Loss_through_AK1-iPSC-NPC.tab\", 'w')\n",
    "rfh.write('Sample\\tCounts\\tType\\tPMD\\n')\n",
    "rfh.flush()\n",
    "'''\n",
    "output table:\n",
    "AK1-iPSC\t#\tGain\tPMD\n",
    "AK1-iPSC\t#\tGain\tnon-PMD\n",
    "AK1-iPSC\t#\tLoss\tNA\n",
    "iPSC-NPC\t#\tGain\tNA\n",
    "iPSC-NPC\t#\tLoss\tNA\n",
    "'''\n",
    "\n",
    "# DMRs in AK1-iPSC relationship\n",
    "dfh = open(\"DMRs_DSS_woPMD_AK1-iPSC_MetTh40.bed\", 'r')\n",
    "gain_nonPMD, loss_nonPMD = 0, 0\n",
    "for i in dfh:\n",
    "\tline = i.strip().split('\\t')\n",
    "\tif float(line[3].split('/')[1].strip(')')) > 0:\n",
    "\t\tloss_nonPMD += 1\n",
    "\telse:\n",
    "\t\tgain_nonPMD += 1\n",
    "dfh.close()\n",
    "\n",
    "dfh = open(\"DMRs_DSS_AK1-iPSC_MetTh40.bed\", 'r')\n",
    "gain_total, loss_total = 0, 0\n",
    "for i in dfh:\n",
    "\tline = i.strip().split('\\t')\n",
    "\tif float(line[3].split('/')[1].strip(')')) > 0:\n",
    "\t\tloss_total += 1\n",
    "\telse:\n",
    "\t\tgain_total += 1\n",
    "\n",
    "rfh.write('AK1-iPSC\\t' + str(gain_total - gain_nonPMD) + '\\tGain\\tPMD\\n')\n",
    "rfh.write('AK1-iPSC\\t' + str(gain_nonPMD) + '\\tGain\\tnon-PMD\\n')\n",
    "rfh.write('AK1-iPSC\\t' + str(loss_nonPMD) + '\\tLoss\\tNA\\n')\n",
    "rfh.flush()\n",
    "\n",
    "# DMRs in iPSC-NPC relationship\n",
    "dfh = open(\"DMRs_DSS_iPSC-NPC_MetTh40.bed\", 'r')\n",
    "gain, loss = 0, 0\n",
    "for i in dfh:\n",
    "\tline = i.strip().split('\\t')\n",
    "\tif float(line[3].split('/')[1].strip(')')) > 0:\n",
    "\t\tloss += 1\n",
    "\telse:\n",
    "\t\tgain += 1\n",
    "dfh.close()\n",
    "\n",
    "print(gain)\n",
    "print(loss)\n",
    "rfh.write('iPSC-NPC\\t' + str(gain) + '\\tGain\\tNA\\n')\n",
    "rfh.write('iPSC-NPC\\t' + str(loss) + '\\tLoss\\tNA\\n')\n",
    "rfh.flush()\n",
    "\n",
    "# /mnt/data/Projects/phenomata/01.Projects/13.AK1_PacBio/01.DNA/Analysis_2023/DMR\n",
    "'''\n",
    "Sample\tCounts\tType\tPMD\n",
    "AK1-iPSC\t305570\tGain\tPMD\n",
    "AK1-iPSC\t79615\tGain\tnon-PMD\n",
    "AK1-iPSC\t3710\tLoss\tNA\n",
    "iPSC-NPC\t4010\tGain\tNA\n",
    "iPSC-NPC\t10422\tLoss\tNA\n",
    "'''\n",
    "gainloss = pd.read_table(\"mCG_Gain_Loss_through_AK1-iPSC-NPC.tab\")\n",
    "gainloss['Type_PMD'] = gainloss.apply(lambda x: f\"{x['Type']} ({x['PMD']})\" if x['Sample'] == 'AK1-iPSC' and x['Type'] == 'Gain' else x['Type'], axis=1)\n",
    "gainloss['Sample_Type'] = gainloss.apply(lambda x: f\"{x['Sample']} {x['Type']}\", axis=1)\n",
    "gainloss_pivot = gainloss.pivot_table(index='Sample_Type', columns='Type_PMD', values='Counts', aggfunc='sum')\n",
    "\n",
    "colors_dic = {'Gain': '#7F0622', 'Gain (PMD)': '#B2697A', 'Gain (non-PMD)': '#7F0622', 'Loss': '#370080'}\n",
    "ax = gainloss_pivot.plot(kind='bar', stacked=True, rot=0, color=colors_dic, figsize=(5,8))\n",
    "ax.set_yticklabels([f'{int(y/10**4)}' for y in ax.get_yticks()])\n",
    "for xticklabel in ax.get_xticklabels():\n",
    "    xticklabel.set_rotation(30)\n",
    "\n",
    "ax.set_ylabel('Number of DMRs (x10,000)')\n",
    "ax.set_xlabel('')\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"DMR_counts_gain-loss.pdf\")\n",
    "\n",
    "# Adding Further Aesthetics through Adobe Illustrator"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Distance to the nearest TSS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/mnt/mone/Project/WC300/Tools/Anaconda3/envs/pygenometracks/bin/python\n",
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import FuncFormatter\n",
    "from matplotlib.ticker import NullLocator\n",
    "sns.set(font=\"arial\", font_scale=1.15, style='ticks')\n",
    "plt.rc(\"axes.spines\", top=False, right=False)\n",
    "\n",
    "os.system('bedtools closest -d -a DMRs_DSS_ALL_MetTh40.merged.bed -b TSS.gencode.v41.annotation.basic.bed > Closest_DMR.bed')\n",
    "os.system('cut -f1-4,9 Closest_DMR.bed > temp_uniq_distance.bed')\n",
    "\n",
    "df = pd.read_table('temp_uniq_distance.bed', header=None, names=['chrom', 'start', 'end', 'DMR', 'Distance'])\n",
    "\n",
    "df['Distance_plus1'] = df['Distance'] + 1\n",
    "mask = df['Distance_plus1'] > 1\n",
    "\n",
    "# Define the number of bins and the range of the data for the non-zero values\n",
    "n_bins = 30\n",
    "data_range = [1, 4000000]\n",
    "\n",
    "# Generate logarithmically spaced bin edges for the non-zero values\n",
    "bin_edges = np.logspace(np.log10(data_range[0]), np.log10(data_range[1]), n_bins + 1)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(7,5), constrained_layout=True)\n",
    "\n",
    "ax.hist(df.loc[mask, 'Distance_plus1'], bins=bin_edges, color='black')\n",
    "ax.set_xscale('log')\n",
    "\n",
    "ax.hist(df.loc[~mask, 'Distance_plus1'], bins=3, color='black')\n",
    "ax.xaxis.set_minor_locator(NullLocator())\n",
    "\n",
    "def format_func(x, pos):\n",
    "    if x == 1:\n",
    "        return '0'\n",
    "    elif x == 10:\n",
    "        return '0.01'\n",
    "    elif x == 100:\n",
    "        return '0.1'\n",
    "    elif x == 1000:\n",
    "        return '1'\n",
    "    elif x == 10000:\n",
    "        return '10'\n",
    "    elif x == 100000:\n",
    "        return '100'\n",
    "    elif x == 1000000:\n",
    "        return '1,000'\n",
    "    else:\n",
    "        return ''\n",
    "\n",
    "ax.xaxis.set_major_formatter(FuncFormatter(format_func))\n",
    "ax.get_yaxis().set_major_formatter(matplotlib.ticker.FuncFormatter(lambda x, p: format(int(x), ',')))\n",
    "ax.set_xlabel(\"Distance to the nearest transcription start site (kb)\")\n",
    "ax.set_ylabel(\"Number of DMR\")\n",
    "\n",
    "plt.savefig(\"DMR_Distance_to_nearest_TSS.png\")\n",
    "os.system(f'/usr/bin/rm temp_uniq_distance.bed')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### K-means Clutering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main Figure (Fig.1 Violinplot)\n",
    "ak1 = pd.read_table('AK1_DMRsmet.tab', index_col=0, header=None)\n",
    "ipsc = pd.read_table('iPSC_DMRsmet.tab', index_col=0, header=None)\n",
    "h1 = pd.read_table('H1_DMRsmet.tab', index_col=0, header=None)\n",
    "hues64 = pd.read_table('HUES64_DMRsmet.tab', index_col=0, header=None)\n",
    "npc = pd.read_table('NPC_DMRsmet.tab', index_col=0, header=None)\n",
    "\n",
    "df_dmrs = pd.concat([ak1, ipsc, h1, hues64, npc], axis=1)\n",
    "df_dmrs.columns.name = 'Sample'\n",
    "df_dmrs.columns = ['AK1', 'iPSC', 'H1', 'HUES64', 'NPC']\n",
    "df_dmrs.index.name = 'Region'\n",
    "\n",
    "df_dmrs = df_dmrs.dropna()\n",
    "\n",
    "def row_scaling(row):\n",
    "    return (row - row.mean()) / row.std()\n",
    "\n",
    "df_dmrs_scaled = df_dmrs.apply(row_scaling, axis=1)\n",
    "\n",
    "within_cluster_ss = list() # Within-cluster Sum of squares\n",
    "BIC = list() # Bayesian Information Criterion (BIC)\n",
    "K = range(2, 21)\n",
    "for k in K:\n",
    "    # Fitting K-means model\n",
    "    kmeans_test = KMeans(n_clusters=k, random_state=np.random.RandomState(seed=42)).fit(df_dmrs_scaled)\n",
    "    # Store Within-cluster Sum of squares\n",
    "    within_cluster_ss.append(kmeans_test.inertia_)\n",
    "    # Fitting Gaussian Mixture Model\n",
    "    gmm_test = GaussianMixture(n_components=k, means_init=kmeans_test.cluster_centers_).fit(df_dmrs_scaled)\n",
    "    # Store BIC\n",
    "    BIC.append(gmm_test.bic(df_dmrs_scaled))\n",
    "    \n",
    "# For Inertia\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 5), constrained_layout=True)\n",
    "sns.lineplot(x=K, y=within_cluster_ss, color='black', ax=ax)\n",
    "ax.set_yticklabels([f'{int(y/10**4)}' for y in ax.get_yticks()])\n",
    "ax.set_ylabel('Within-Cluster Sum of Squares (x10,000)')\n",
    "ax.set_xlabel('Number of K')\n",
    "\n",
    "# For BIC\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 5), constrained_layout=True)\n",
    "sns.lineplot(x=K, y=BIC, color='black', ax=ax)\n",
    "ax.set_yticklabels([f'{float(y/10**6)}' for y in ax.get_yticks()])\n",
    "ax.set_ylabel('Bayesian Information Criterion (x1,000,000)')\n",
    "ax.set_xlabel('Number of K')\n",
    "\n",
    "# Combined\n",
    "fig, axes = plt.subplots(1, 2, figsize=(9, 5), constrained_layout=True)\n",
    "sns.lineplot(x=K, y=within_cluster_ss, color='black', ax=axes[0])\n",
    "axes[0].set_yticklabels([f'{int(y/10**4)}' for y in axes[0].get_yticks()])\n",
    "axes[0].set_ylabel('Within-Cluster Sum of Squares (x10,000)')\n",
    "axes[0].set_xlabel('Number of K')\n",
    "sns.lineplot(x=K, y=BIC, color='black', ax=axes[1])\n",
    "axes[1].set_yticklabels([f'{float(y/10**6)}' for y in axes[1].get_yticks()])\n",
    "axes[1].set_ylabel('Bayesian Information Criterion (x1,000,000)')\n",
    "axes[1].set_xlabel('Number of K')\n",
    "\n",
    "\n",
    "#for xticklabel in ax.get_xticklabels():\n",
    "#    xticklabel.set_rotation(30)\n",
    "\n",
    "K = 6\n",
    "kmeans_K6 = KMeans(n_clusters=K, random_state=np.random.RandomState(seed=42)).fit(df_dmrs_scaled)\n",
    "df_dmrs[f'K-means cluster (K={K})'] = kmeans_K6.labels_\n",
    "\n",
    "df_dmrs.to_csv(\"DMRs_Kmeans_K6.tab\", sep='\\t')\n",
    "\n",
    "'''\n",
    "df_dmrs_melted = pd.melt(df_dmrs, id_vars='K-means cluster (K=12)', var_name='variable', value_name='value')\n",
    "\n",
    "fig, axes = plt.subplots(1, 1, figsize=(8, 5), constrained_layout=True)\n",
    "sns.heatmap(df_dmrs_melted.pivot_table(index='K-means cluster (K=12)', columns='variable', values='value'),\n",
    "            cmap=cmap)\n",
    "\n",
    "sns.heatmap(df_dmrs.sort_values(by='K-means cluster (K=12)').iloc[:, :5], xticklabels=True, yticklabels=False, vmin=0, vmax=100, cmap=cmap)\n",
    "plt.tight_layout()\n",
    "'''\n",
    "\n",
    "K = 6\n",
    "K_range = range(0, K)\n",
    "df_dmrs_dic = dict()\n",
    "for k in K_range:\n",
    "    df_names = f'K{k}'\n",
    "    temp_df = df_dmrs[df_dmrs[f'K-means cluster (K={K})'] == k]\n",
    "    temp_df_filtered = temp_df.iloc[:, :5]\n",
    "    temp_df_filtered_reset = temp_df_filtered.reset_index()\n",
    "    temp_df_melted = pd.melt(temp_df_filtered_reset, id_vars='Region', var_name='Sample', value_name='DNA methylation (%)')\n",
    "    df_dmrs_dic[df_names] = temp_df_melted\n",
    "\n",
    "\n",
    "df_dmrs_list = [df_dmrs_dic[f'K{k}'] for k in K_range]\n",
    "df_dmrs_titles = [f'C{k+1} (n={format(len(np.unique(df_dmrs_dic[f\"K{k}\"][\"Region\"].values)), \",\")})' for k in K_range] # the number of DMRs for each K cluster: np.unique(df_dmrs_dic['K11']['Region'].values)\n",
    "\n",
    "\n",
    "fig, axes = plt.subplots(3, 2, figsize=(11, 18), constrained_layout=True)\n",
    "for data, title, ax in zip(df_dmrs_list, df_dmrs_titles, axes.flat):\n",
    "    \n",
    "    sns.pointplot(data=data, x='Sample', y='DNA methylation (%)', color='black', errorbar=\"sd\", scale=0.5, ax=ax)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_title(title, fontweight='bold')\n",
    "    ax.set_ylim(0,100)\n",
    "\n",
    "plt.savefig('DMR_K-means_clusters_C1-C6.png')\n",
    "\n",
    "\n",
    "'''\n",
    "df_dmrs_k0 = df_dmrs[df_dmrs['K-means cluster (K=12)'] == 0]\n",
    "df_dmrs_k0_filtered = df_dmrs_k0.iloc[:, :5]\n",
    "df_dmrs_k0_filtered_reset = df_dmrs_k0_filtered.reset_index()\n",
    "df_dmrs_k0_melted = pd.melt(df_dmrs_k0_filtered_reset, id_vars='Region', var_name='Sample', value_name='DNA methylation (%)')\n",
    "catplot = sns.catplot(data=df_dmrs_k0_melted, x='Sample', y='DNA methylation (%)', kind='point', estimator='median', errorbar=('ci', 95), color='black', height=7, aspect=1.2)\n",
    "catplot.ax.set_ylim(0,100)\n",
    "plt.savefig('test_DMR_K-means_cluster0_mCG.png')\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Ontology Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "K=6\n",
    "K_range = range(0, K)\n",
    "df_dmrs = pd.read_table(f\"DMRs_Kmeans_K{6}.tab\", index_col=0)\n",
    "closest_tss = dict()\n",
    "\n",
    "with open(\"Closest_DMR.bed\", 'r') as dfh:\n",
    "    for i in dfh:\n",
    "        line = i.strip('\\n').split('\\t')\n",
    "        closest_tss[line[0] + ':' + line[1] + '-' + line[2]] = line[3] + '/' + line[7]\n",
    "        \n",
    "        \n",
    "def get_gene_name_from_index(dataframe):\n",
    "    genelist = list()\n",
    "    for index in dataframe.index:\n",
    "        if closest_tss[index].split('/')[3] not in genelist:\n",
    "            genelist.append(closest_tss[index].split('/')[3])\n",
    "    return genelist\n",
    "        \n",
    "gobp = gp.get_library(name='GO_Biological_Process_2023', organism='Human')\n",
    "kegg = gp.get_library(name='KEGG_2021_Human', organism='Human')\n",
    "background_genes = get_gene_name_from_index(df_dmrs)\n",
    "\n",
    "for k in K_range:\n",
    "    observed_genes = df_dmrs[f'K-means cluster (K={K})' == k+1]\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "df_dmrs[f'K-means cluster (K={K})']\n",
    "df_dmrs[df_dmrs[f'K-means cluster (K={K})'] == 1]\n",
    "\n",
    "\n",
    "\n",
    "k1_genes = get_gene_name_from_index(df_dmrs[df_dmrs[f'K-means cluster (K={K})'] == 1])\n",
    "enr_go = gp.enrichr(gene_list=k1_genes,\n",
    "                 gene_sets=gobp,\n",
    "                 background=background_genes,\n",
    "                 outdir=None)\n",
    "                  \n",
    "enr_kegg = gp.enrichr(gene_list=k1_genes,\n",
    "                 gene_sets=kegg,\n",
    "                 background=background_genes,\n",
    "                 outdir=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
